diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index a27f489..5e82fc0 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -353,6 +353,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6dl-sabrelite.dtb \
 	imx6dl-sabresd.dtb \
 	imx6dl-sabresd-btwifi.dtb \
+	imx6dl-sabresd-btwifi-m2.dtb \
+	imx6dl-sabresd-btwifi-m2-oob.dtb \
 	imx6dl-sabresd-hdcp.dtb \
 	imx6dl-sabresd-ldo.dtb \
 	imx6dl-ts4900.dtb \
@@ -418,6 +420,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6q-sabrelite.dtb \
 	imx6q-sabresd.dtb \
 	imx6q-sabresd-btwifi.dtb \
+	imx6q-sabresd-btwifi-m2.dtb \
+	imx6q-sabresd-btwifi-m2-oob.dtb \
 	imx6q-sabresd-hdcp.dtb \
 	imx6q-sabresd-ldo.dtb \
 	imx6q-sabresd-enetirq.dtb \
@@ -441,6 +445,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6qp-sabreauto-flexcan1.dtb \
 	imx6qp-sabreauto-gpmi-weim.dtb \
 	imx6qp-sabresd-btwifi.dtb \
+	imx6qp-sabresd-btwifi-m2.dtb \
+	imx6qp-sabresd-btwifi-m2-oob.dtb \
 	imx6qp-sabresd-hdcp.dtb \
 	imx6qp-sabresd-ldo.dtb \
 	imx6qp-sabresd-ldo-pcie-cert.dtb \
@@ -448,6 +454,7 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 dtb-$(CONFIG_SOC_IMX6SL) += \
 	imx6sl-evk.dtb \
 	imx6sl-evk-btwifi.dtb \
+	imx6sl-evk-btwifi-m2.dtb \
 	imx6sl-evk-ldo.dtb \
 	imx6sl-evk-csi.dtb \
 	imx6sl-evk-uart.dtb \
@@ -462,6 +469,8 @@ dtb-$(CONFIG_SOC_IMX6SX) += \
 	imx6sx-sdb-reva-ldo.dtb \
 	imx6sx-sdb.dtb	\
 	imx6sx-sdb-btwifi.dtb \
+	imx6sx-sdb-btwifi-m2.dtb \
+	imx6sx-sdb-btwifi-m2-oob.dtb \
 	imx6sx-sdb-emmc.dtb \
 	imx6sx-sdb-lcdif1.dtb \
 	imx6sx-sdb-ldo.dtb	\
@@ -484,6 +493,10 @@ dtb-$(CONFIG_SOC_IMX6UL) += \
 	imx6ul-14x14-evk.dtb \
 	imx6ul-14x14-evk-btwifi.dtb \
 	imx6ul-14x14-evk-btwifi-oob.dtb \
+	imx6ul-14x14-evk-btwifi-m2.dtb \
+	imx6ul-14x14-evk-btwifi-m2-oob.dtb \
+	imx6ul-14x14-evk-btwifi-m2-uhs.dtb \
+	imx6ul-14x14-evk-btwifi-m2-oob-uhs.dtb \
 	imx6ul-14x14-evk-csi.dtb \
 	imx6ul-14x14-evk-emmc.dtb \
 	imx6ul-14x14-evk-gpmi-weim.dtb \
@@ -527,6 +540,10 @@ dtb-$(CONFIG_SOC_IMX6ULL) += \
 	imx6ull-14x14-evk.dtb \
 	imx6ull-14x14-evk-btwifi.dtb \
 	imx6ull-14x14-evk-btwifi-oob.dtb \
+	imx6ull-14x14-evk-btwifi-m2.dtb \
+	imx6ull-14x14-evk-btwifi-m2-oob.dtb \
+	imx6ull-14x14-evk-btwifi-m2-uhs.dtb \
+	imx6ull-14x14-evk-btwifi-m2-oob-uhs.dtb \
 	imx6ull-14x14-evk-emmc.dtb \
 	imx6ull-14x14-evk-gpmi-weim.dtb \
 	imx6ull-14x14-evk-usb-certi.dtb \
diff --git a/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..602f401
--- /dev/null
+++ b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6dl-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..4742495
--- /dev/null
+++ b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6dl-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6dqscm-qwks-rev3-btwifi.dtsi b/arch/arm/boot/dts/imx6dqscm-qwks-rev3-btwifi.dtsi
index dc41db6..e096ca2 100644
--- a/arch/arm/boot/dts/imx6dqscm-qwks-rev3-btwifi.dtsi
+++ b/arch/arm/boot/dts/imx6dqscm-qwks-rev3-btwifi.dtsi
@@ -4,6 +4,8 @@
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 as
  * published by the Free Software Foundation.
+ * 
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration. 
  */
 
 / {
@@ -14,22 +16,9 @@
 		#reset-cells = <0>;
 	};
 
-	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio6 5 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
-	};
-
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		wlreg_on-supply = <&wlreg_on>;
-		gpios = <&gpio6 4 0>; /* WL_HOST_WAKE */
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio6 5 GPIO_ACTIVE_LOW>;
 	};
 };
 
@@ -43,14 +32,25 @@
 };
 
 &usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_wifi>;
 	bus-width = <4>;
 	no-1-8-v;
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio6>;
+		interrupts = <4 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO6_IO4 active high. Change to IRQ_TYPE_EDGE_RISING for 1BW and SN8000. */
+		interrupt-names = "host-wake";
+	};
 };
 
 &iomuxc {
@@ -70,8 +70,8 @@
 				MX6QDL_PAD_SD2_DAT1__SD2_DATA1		0x17059
 				MX6QDL_PAD_SD2_DAT2__SD2_DATA2		0x17059
 				MX6QDL_PAD_SD2_DAT3__SD2_DATA3		0x17059
-				MX6QDL_PAD_CSI0_DAT19__GPIO6_IO05	0x13069
-				MX6QDL_PAD_CSI0_DAT18__GPIO6_IO04	0x13069
+				MX6QDL_PAD_CSI0_DAT19__GPIO6_IO05	0x13069  /* WL_REG_ON */
+				MX6QDL_PAD_CSI0_DAT18__GPIO6_IO04	0x13069  /* WL_HOST_WAKE */
 			>;
 		};
 		pinctrl_uart4: uart4grp {
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..2e8b621
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6q-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..48845c8
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi
@@ -0,0 +1,16 @@
+&iomuxc {
+	imx6qdl-sabresd-murata-v2 {
+		/* add MUXing entry for SD2 4-bit interface and configure control pins */
+		pinctrl_wifi_oob: pinctrl_wifi {
+			fsl,pins = <
+				MX6QDL_PAD_KEY_COL0__GPIO4_IO06		0x0b000	/* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+
+&brcmf{
+	interrupt-parent = <&gpio4>;
+	interrupts = <6 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..7985b59
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6q-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi
new file mode 100644
index 00000000..08d9098
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/*
+ * NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into SD2
+ * slot using Murata i.MX InterConnect Ver 2.0 Adapter & connecting Bluetooth
+ * UART & control signals via ribbon cable.
+ * This configuration supports both WLAN and Bluetooth.
+ * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected via ribbon cable (J13 connector
+ * on board).
+ * ==> Hardware modification is required. Refer to schematic.
+ *
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration.
+ */
+
+/ {
+	leds {
+		compatible = "gpio-leds";
+		status = "disabled";
+	};
+
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio1 2 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio4 7 GPIO_ACTIVE_LOW>;
+ 	};
+};
+
+&ecspi1 {
+	status = "disabled";
+};
+
+&iomuxc {
+	imx6qdl-sabresd-murata-v2 {
+		pinctrl_btreg: btreggrp {
+			fsl,pins = <
+				MX6QDL_PAD_GPIO_2__GPIO1_IO02 0x1b0b0  /* BT_REG_ON */
+			>;
+		};
+
+		/* add MUXing entry for SD2 4-bit interface and configure control pins */
+		pinctrl_wifi: wifigrp {
+			fsl,pins = <
+				MX6QDL_PAD_SD2_CMD__SD2_CMD		0x17059
+				MX6QDL_PAD_SD2_CLK__SD2_CLK		0x10059
+				MX6QDL_PAD_SD2_DAT0__SD2_DATA0		0x17059
+				MX6QDL_PAD_SD2_DAT1__SD2_DATA1		0x17059
+				MX6QDL_PAD_SD2_DAT2__SD2_DATA2		0x17059
+				MX6QDL_PAD_SD2_DAT3__SD2_DATA3		0x17059
+				MX6QDL_PAD_KEY_ROW0__GPIO4_IO07		0x13069	/* WL_REG_ON */
+				MX6QDL_PAD_KEY_COL0__GPIO4_IO06		0x13069	/* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+
+/* re-defining gpio_leds removes conflict with BT_REG_ON */
+&pinctrl_gpio_leds {
+	fsl,pins = <
+	>;
+};
+
+&uart5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart5_1
+		     &pinctrl_btreg>;
+	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode; */
+	/* pinctrl-0 = <&pinctrl_uart5dte_1>; */
+};
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_wifi>;
+	bus-width = <4>;
+	no-1-8-v;
+	non-removable;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm/boot/dts/imx6qdl-sabresd-btwifi.dtsi b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi.dtsi
index 7164a9e..93a79e5 100644
--- a/arch/arm/boot/dts/imx6qdl-sabresd-btwifi.dtsi
+++ b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi.dtsi
@@ -14,6 +14,8 @@
  * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected via ribbon cable (J13 connector
  * on board).
  * ==> Hardware modification is required. Refer to schematic.
+ *
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration. 
  */
 
 / {
@@ -29,22 +31,9 @@
 		#reset-cells = <0>;
 	};
 
-	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio4 7 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
-	};
-
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		gpios = <&gpio4 6 0>; /* WL_HOST_WAKE */
-		wlreg_on-supply = <&wlreg_on>;
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio4 7 GPIO_ACTIVE_LOW>;
 	};
 };
 
@@ -56,7 +45,7 @@
 	imx6qdl-sabresd-murata-v2 {
 		pinctrl_btreg: btreggrp {
 			fsl,pins = <
-				MX6QDL_PAD_GPIO_2__GPIO1_IO02 0x1b0b0
+				MX6QDL_PAD_GPIO_2__GPIO1_IO02 0x1b0b0  /* BT_REG_ON */
 			>;
 		};
 
@@ -94,12 +83,23 @@
 };
 
 &usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_wifi>;
 	bus-width = <4>;
 	no-1-8-v;
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio4>;
+		interrupts = <6 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO4_IO6 active high. Change to IRQ_TYPE_EDGE_RISING for 1BW and SN8000. */
+		interrupt-names = "host-wake";
+	};
 };
diff --git a/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..b08baf8
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6qp-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..1c4a097
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6qp-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts
new file mode 100644
index 00000000..d3d1df2
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/* NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into SD1
+ * slot using Murata i.MX InterConnect Ver 1.0 Adapter AND wiring in control
+ * signals with SD Card Extender on SD3 slot.
+ * Bluetooth UART connect via SD1 EMMC/MMC Plus pinout.
+ * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected from SD Card Extender on SD3
+ * slot.
+ *
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration.
+ */
+#include "imx6sl-evk.dts"
+
+/ {
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 17 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 16 GPIO_ACTIVE_LOW>;
+ 	};
+};
+
+&iomuxc {
+	imx6sl-evk-murata-v1_sdext {
+		/* Only MUX SD1_DAT0..3 lines so UART4 can be MUXed on higher data lines. */
+		pinctrl_btreg: btreggrp {
+			fsl,pins = <
+				MX6SL_PAD_SD3_DAT3__GPIO5_IO17          0x13069 /* BT_REG_ON */
+			>;
+		};
+		pinctrl_wifi: wifigrp {
+			fsl,pins = <
+				MX6SL_PAD_SD1_CMD__SD1_CMD		0x17059
+				MX6SL_PAD_SD1_CLK__SD1_CLK		0x10059
+				MX6SL_PAD_SD1_DAT0__SD1_DATA0		0x17059
+				MX6SL_PAD_SD1_DAT1__SD1_DATA1		0x17059
+				MX6SL_PAD_SD1_DAT2__SD1_DATA2		0x17059
+				MX6SL_PAD_SD1_DAT3__SD1_DATA3		0x17059
+				MX6SL_PAD_SD3_DAT1__GPIO5_IO20		0x13069 /* WL_HOST_WAKE */
+				MX6SL_PAD_SD3_DAT2__GPIO5_IO16		0x13069 /* WL_REG_ON */
+			>;
+		};
+
+		pinctrl_usdhc3_1: usdhc3grp-1 {
+			fsl,pins = <
+				MX6SL_PAD_SD3_CMD__SD3_CMD		0x17059
+				MX6SL_PAD_SD3_CLK__SD3_CLK		0x10059
+				MX6SL_PAD_SD3_DAT0__SD3_DATA0		0x17059
+			>;
+		};
+	};
+};
+
+/* Murata: declare UART4 interface for Bluetooth. */
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart4_1
+		     &pinctrl_btreg>;
+	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode; */
+	/* pinctrl-0 = <&pinctrl_uart4dte_1>; */
+};
+
+&usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_wifi>;
+	bus-width = <4>;
+	no-1-8-v;
+	non-removable;
+	cap-power-off-card;
+	pm-ignore-notify;
+	wifi-host;
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
+
+&usdhc3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usdhc3_1>;
+	bus-width = <1>;
+	no-1-8-v;
+};
diff --git a/arch/arm/boot/dts/imx6sl-evk-btwifi.dts b/arch/arm/boot/dts/imx6sl-evk-btwifi.dts
index 4eac83e..3f1a584 100644
--- a/arch/arm/boot/dts/imx6sl-evk-btwifi.dts
+++ b/arch/arm/boot/dts/imx6sl-evk-btwifi.dts
@@ -12,6 +12,8 @@
  * Bluetooth UART connect via SD1 EMMC/MMC Plus pinout.
  * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected from SD Card Extender on SD3
  * slot.
+ * 
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration. 
  */
 #include "imx6sl-evk.dts"
 
@@ -23,22 +25,9 @@
 		#reset-cells = <0>;
 	};
 
-	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio5 16 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
-	};
-
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		gpios = <&gpio5 20 0>; /* WL_HOST_WAKE */
-		wlreg_on-supply = <&wlreg_on>;
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 16 GPIO_ACTIVE_LOW>;
 	};
 };
 
@@ -73,6 +62,7 @@
 		};
 	};
 };
+
 /* Murata: declare UART4 interface for Bluetooth. */
 &uart4 {
 	pinctrl-names = "default";
@@ -87,14 +77,25 @@
 };
 
 &usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_wifi>;
 	bus-width = <4>;
 	no-1-8-v;
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio5>;
+		interrupts = <20 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO5_IO20 active high. Change to IRQ_TYPE_EDGE_RISING for 1BW and SN8000. */
+		interrupt-names = "host-wake";
+	};
 };
 
 &usdhc3 {
diff --git a/arch/arm/boot/dts/imx6sll-evk-btwifi.dts b/arch/arm/boot/dts/imx6sll-evk-btwifi.dts
index db30b1f..65312ae 100644
--- a/arch/arm/boot/dts/imx6sll-evk-btwifi.dts
+++ b/arch/arm/boot/dts/imx6sll-evk-btwifi.dts
@@ -10,6 +10,8 @@
  * NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into Slot
  * SD1 and using Murata i.MX InterConnect Ver 2.0 Adapter. Bluetooth UART &
  * control signals are connected via ribbon cable (J4 connector).
+ * 
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration. 
  */
 
 #include "imx6sll-evk.dts"
@@ -22,22 +24,9 @@
 		#reset-cells = <0>;
 	};
 
-	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio3 24 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
-	};
-
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		wlreg_on-supply = <&wlreg_on>;
-		gpios = <&gpio3 26 0>;          /* WL_HOST_WAKE */
+	usdhc3_pwrseq: usdhc3_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio3 24 GPIO_ACTIVE_LOW>;
 	};
 };
 
@@ -73,11 +62,22 @@
 };
 
 &usdhc3 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_wifi>;
 	no-1-8-v;
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host; /* add hook for SD card detect mechanism for BCMDHD driver */
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc3_pwrseq>;
+	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio3>;
+		interrupts = <26 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO3_IO26 active high. Change to IRQ_TYPE_EDGE_RISING for 1BW and SN8000. */
+		interrupt-names = "host-wake";
+	};
 };
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts
new file mode 100644
index 00000000..0a19a8d
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6sx-sdb-btwifi-m2.dts"
+#include "imx6sx-sdb-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..452bbef
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi
@@ -0,0 +1,18 @@
+&iomuxc {
+	imx6sx-sdb-murata-v1_sdext {
+		pinctrl_wifi_oob: pinctrl_wifi {
+			fsl,pins = <
+				// Murata Module control signals -- not connected (R327 and need to disable ENET1 CLK)
+				MX6SX_PAD_ENET1_RX_CLK__GPIO2_IO_4	0x0b000 /* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+&fec1 {
+	status = "disabled";
+};
+&brcmf{
+	interrupt-parent = <&gpio2>;
+	interrupts = <4 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts
new file mode 100644
index 00000000..3038da0
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6sx-sdb.dts"
+#include "imx6sx-sdb-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi
new file mode 100644
index 00000000..4c7d8d7
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/*
+ * NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into SD3
+ * slot using Murata i.MX InterConnect Ver 1.0 Adapter AND SD Card Extender on
+ * SD2 slot. Bluetooth UART connects via SD3 EMMC/MMC Plus pinout.
+ * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE connect via SD Card Extender.
+ */
+
+/ {
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio4 10 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+//		reset-gpios = <&gpio2 4 GPIO_ACTIVE_LOW>;  /* currently not implemented */
+	};
+};
+
+&iomuxc {
+	imx6sx-sdb-murata-wifibt {
+		pinctrl_bt: btgrp {
+			fsl,pins = <
+				MX6SX_PAD_NAND_DATA06__GPIO4_IO_10	0x13069 /* BT_REG_ON */
+			>;
+		};
+
+		pinctrl_wifi: wifigrp {
+			fsl,pins = <
+				/* Murata Module control signals -- not connected (R327 and need to disable ENET1 CLK) */
+//				MX6SX_PAD_ENET1_RX_CLK__GPIO2_IO_4	0x13069 /* WL_REG_ON */
+			>;
+		};
+	};
+};
+
+&uart5 { /* for bluetooth */
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart5 &pinctrl_bt>;
+	uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode;*/
+	/* pinctrl-0 = <&pinctrl_uart5dte_1>; */
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usdhc2 &pinctrl_wifi>;
+	bus-width = <4>;
+	no-1-8-v;	/* force 3.3V VIO */
+	non-removable;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	cap-power-off-card;
+	status = "okay";
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi.dts b/arch/arm/boot/dts/imx6sx-sdb-btwifi.dts
index 698e49e..8dc07bf 100644
--- a/arch/arm/boot/dts/imx6sx-sdb-btwifi.dts
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi.dts
@@ -11,6 +11,8 @@
  * slot using Murata i.MX InterConnect Ver 1.0 Adapter AND SD Card Extender on
  * SD2 slot. Bluetooth UART connects via SD3 EMMC/MMC Plus pinout.
  * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE connect via SD Card Extender.
+ * 
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration.
  */
 
 #include "imx6sx-sdb.dts"
@@ -23,23 +25,9 @@
 		#reset-cells = <0>;
 	};
 
-	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio6 10 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
-	};
-
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		/* WL_HOST_WAKE: SD2_DAT1 (gpio6 9) */
-		gpios = <&gpio6 9 0>;
-		wlreg_on-supply = <&wlreg_on>;
+	usdhc3_pwrseq: usdhc3_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio6 10 GPIO_ACTIVE_LOW>;
 	};
 };
 
@@ -108,13 +96,23 @@
 };
 
 &usdhc3 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_wifi>;
 	bus-width = <4>;
 	no-1-8-v;	/* force 3.3V VIO */
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host;	/* pull in card detect mechanism for BCMDHD driver */
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc3_pwrseq>;
 	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio6>;
+		interrupts = <9 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO6_IO9 active high. Change to IRQ_TYPE_EDGE_RISING for 1BW and SN8000. */
+		interrupt-names = "host-wake";
+	};
 };
diff --git a/arch/arm/boot/dts/imx6sxscm-evb-btwifi.dtsi b/arch/arm/boot/dts/imx6sxscm-evb-btwifi.dtsi
index 3658389..2bd3d79 100644
--- a/arch/arm/boot/dts/imx6sxscm-evb-btwifi.dtsi
+++ b/arch/arm/boot/dts/imx6sxscm-evb-btwifi.dtsi
@@ -4,6 +4,8 @@
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 as
  * published by the Free Software Foundation.
+ * 
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration. 
  */
 
 / {
@@ -14,26 +16,15 @@
 		#reset-cells = <0>;
 	};
 
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio4 8 GPIO_ACTIVE_LOW>;
+	};
+
 	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio4 8 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
 		vcc_sd3: regulator@0 {
 			status = "disabled";
 		};
-
-	};
-
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		wlreg_on-supply = <&wlreg_on>;
-		gpios = <&gpio7 9 0>; /* WL_HOST_WAKE */
 	};
 };
 
@@ -97,8 +88,8 @@
 				MX6SX_PAD_SD2_DATA2__USDHC2_DATA2	0x17069
 				MX6SX_PAD_SD2_DATA3__USDHC2_DATA3	0x17069
 				/* Murata Module control signals */
-				MX6SX_PAD_NAND_DATA04__GPIO4_IO_8	0x13069
-				MX6SX_PAD_SD3_DATA7__GPIO7_IO_9         0x13069
+				MX6SX_PAD_NAND_DATA04__GPIO4_IO_8	0x13069  /* WL_REG_ON */
+				MX6SX_PAD_SD3_DATA7__GPIO7_IO_9         0x13069  /* WL_HOST_WAKE */
 			>;
 		};
 	};
@@ -114,11 +105,23 @@
 };
 
 &usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-0 = <&pinctrl_wifi>;
 	bus-width = <4>;
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio7>;
+		interrupts = <9 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO7_IO9 active high */
+		interrupt-names = "host-wake";
+	};
 };
 
 &usdhc3 {
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts
new file mode 100644
index 00000000..c7a4c1a
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts
@@ -0,0 +1,21 @@
+#include "imx6ul-14x14-evk-btwifi-m2-uhs.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
+
+&fec2 {
+	status="disabled";
+};
+
+&fec1 {
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+			micrel,led-mode = <1>;
+			clocks = <&clks IMX6UL_CLK_ENET_REF>;
+			clock-names = "rmii-ref";
+		};
+	};
+};
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts
new file mode 100644
index 00000000..5826e96
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6ul-14x14-evk-btwifi-m2.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts
new file mode 100644
index 00000000..4d16874
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts
@@ -0,0 +1,10 @@
+#include "imx6ul-14x14-evk-btwifi-m2.dts"
+
+&usdhc1 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
+	/delete-property/ no-1-8-v;
+};
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts
new file mode 100644
index 00000000..884e1bd
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts
@@ -0,0 +1,43 @@
+#include "imx6ul-14x14-evk-btwifi.dts"
+
+/ {
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 1 GPIO_ACTIVE_LOW>; /* WL REG ON */
+		post-power-on-delay-ms = <120>;  /* Murata - add delay before accessing SDIO bus */
+	};
+};
+
+&wlreg_on {
+	status = "disabled";
+};
+
+&bcmdhd_wlan_0 {
+	status = "disabled";
+};
+
+&pinctrl_usdhc1 {
+        fsl,pins = <
+            	MX6UL_PAD_GPIO1_IO05__GPIO1_IO05	0x17059 /* SD1 VSELECT */
+        >;
+};
+
+&usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	fsl,tuning-start-tap = <20>;
+	fsl,tuning-step = <2>;
+	non-removable;
+	/delete-property/ cd-post;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	/delete-property/ wifi-host;
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk.dts b/arch/arm/boot/dts/imx6ul-14x14-evk.dts
index db8427f..cc1af56 100644
--- a/arch/arm/boot/dts/imx6ul-14x14-evk.dts
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk.dts
@@ -204,12 +204,26 @@
 	};
 };
 
+/* Murata: need to redefine FEC1 and FEC2 constructs to keep MDIO declarations independent. 
+ * Needed by OOB IRQ functionality - see imx6ul-evk-btwifi-oob.dtsi 
+ */
+
 &fec1 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_enet1>;
 	phy-mode = "rmii";
 	phy-handle = <&ethphy0>;
 	status = "okay";
+
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+		};
+	};
 };
 
 &fec2 {
@@ -223,11 +237,6 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		ethphy0: ethernet-phy@2 {
-			compatible = "ethernet-phy-ieee802.3-c22";
-			reg = <2>;
-		};
-
 		ethphy1: ethernet-phy@1 {
 			compatible = "ethernet-phy-ieee802.3-c22";
 			reg = <1>;
diff --git a/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..66e55dd
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2017 NXP
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+&pinctrl_wifi {
+	fsl,pins = <
+		/* MUXing for WL_HOST_WAKE */
+		MX6UL_PAD_ENET2_RX_ER__GPIO2_IO15  0x0b000  /* input: 100K Pull-up */
+	>;
+};
+
+/*
+ * For WL_HOST_WAKE (OOB_IRQ) to function correctly, we must disable
+ * the secondary ethernet port (FEC2). Hardware re-work is to remove
+ * R1633 and populate R1704 with 0 Ohm resistor.
+ * Refer to Murata Hardware Reference Manual for more details.
+ */
+&fec2 {
+	status = "disabled";
+};
+
+&brcmf {
+	interrupt-parent = <&gpio2>;
+	interrupts = <15 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts
new file mode 100644
index 00000000..8da2a49
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts
@@ -0,0 +1,21 @@
+#include "imx6ull-14x14-evk-btwifi-m2-uhs.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
+
+&fec2 {
+	status="disabled";
+};
+
+&fec1 {
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+			micrel,led-mode = <1>;
+			clocks = <&clks IMX6UL_CLK_ENET_REF>;
+			clock-names = "rmii-ref";
+		};
+	};
+};
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts
new file mode 100644
index 00000000..618f222
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6ull-14x14-evk-btwifi-m2.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts
new file mode 100644
index 00000000..8b4112c
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts
@@ -0,0 +1,10 @@
+#include "imx6ull-14x14-evk-btwifi-m2.dts"
+
+&usdhc1 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
+	/delete-property/ no-1-8-v;
+};
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts
new file mode 100644
index 00000000..52fbe0f
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts
@@ -0,0 +1,41 @@
+#include "imx6ull-14x14-evk-btwifi.dts"
+
+/ {
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 1 GPIO_ACTIVE_LOW>; /* WL REG ON */
+		post-power-on-delay-ms = <120>;  /* Murata - add delay before accessing SDIO bus */
+	};
+};
+
+&wlreg_on {
+	status = "disabled";
+};
+
+&bcmdhd_wlan_0 {
+	status = "disabled";
+};
+
+&pinctrl_hog_1 {
+        fsl,pins = <
+            	MX6UL_PAD_GPIO1_IO05__GPIO1_IO05	0x17059 /* SD1 VSELECT */
+        >;
+};
+
+&usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	non-removable;
+	/delete-property/ cd-post;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	/delete-property/ wifi-host;
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk.dts b/arch/arm/boot/dts/imx6ull-14x14-evk.dts
index 143250b..0a91d04 100644
--- a/arch/arm/boot/dts/imx6ull-14x14-evk.dts
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk.dts
@@ -167,12 +167,26 @@
 	};
 };
 
+/* Murata: need to redefine FEC1 and FEC2 constructs to keep MDIO declarations independent. 
+ * Needed by OOB IRQ functionality - see imx6ul-evk-btwifi-oob.dtsi 
+ */
+
 &fec1 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_enet1>;
 	phy-mode = "rmii";
 	phy-handle = <&ethphy0>;
 	status = "okay";
+
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+		};
+	};
 };
 
 &fec2 {
@@ -186,11 +200,6 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		ethphy0: ethernet-phy@2 {
-			compatible = "ethernet-phy-ieee802.3-c22";
-			reg = <2>;
-		};
-
 		ethphy1: ethernet-phy@1 {
 			compatible = "ethernet-phy-ieee802.3-c22";
 			reg = <1>;
diff --git a/arch/arm/boot/dts/imx7d-sdb.dts b/arch/arm/boot/dts/imx7d-sdb.dts
index 8445a25..eb89637 100644
--- a/arch/arm/boot/dts/imx7d-sdb.dts
+++ b/arch/arm/boot/dts/imx7d-sdb.dts
@@ -38,6 +38,8 @@
  *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  *     OTHER DEALINGS IN THE SOFTWARE.
+ * 
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ. 
  */
 
 /dts-v1/;
@@ -63,6 +65,11 @@
 		#reset-cells = <0>;
 	};
 
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio4 21 GPIO_ACTIVE_LOW>;
+	};
+
 	regulators {
 		compatible = "simple-bus";
 		#address-cells = <1>;
@@ -115,16 +122,6 @@
 			off-on-delay = <20000>;
 			enable-active-high;
 		};
-
-		wlreg_on: fixedregulator@6 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio4 21 GPIO_ACTIVE_HIGH>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
 	};
 
 	backlight {
@@ -135,12 +132,6 @@
 		status = "okay";
 	};
 
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		gpios = <&gpio4 20 0>;		/* WL_HOST_WAKE */
-		wlreg_on-supply = <&wlreg_on>;
-	};
-
 	pxp_v4l2_out {
 		compatible = "fsl,imx7d-pxp-v4l2", "fsl,imx6sx-pxp-v4l2", "fsl,imx6sl-pxp-v4l2";
 		status = "okay";
@@ -624,6 +615,10 @@
 	};
 };
 
+&pcie_phy{
+       status = "okay";
+};
+
 &pcie {
 	pinctrl-names = "default";
 	reset-gpio = <&gpio_spi 1 GPIO_ACTIVE_LOW>;
@@ -729,7 +724,7 @@
 	/* for DTE mode, add below change */
 	/* fsl,dte-mode; */
 	/* pinctrl-0 = <&pinctrl_uart5dte>; */
-	status = "okay";
+	status = "disabled";
 };
 
 &uart6 {
@@ -771,17 +766,25 @@
 };
 
 &usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default", "state_100mhz", "state_200mhz";
 	pinctrl-0 = <&pinctrl_usdhc2>;
 	pinctrl-1 = <&pinctrl_usdhc2_100mhz>;
 	pinctrl-2 = <&pinctrl_usdhc2_200mhz>;
-	enable-sdio-wakeup;
-	keep-power-in-suspend;
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc2_pwrseq>;
 	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio4>;
+		interrupts = <20 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO4_IO20 active high */
+		interrupt-names = "host-wake";
+	};
 };
 
 &usdhc3 {
@@ -805,7 +808,7 @@
 &gpmi {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_gpmi_nand_1>;
-	status = "disabled";
+	status = "okay";
 	nand-on-flash-bbt;
 };
 
@@ -1175,6 +1178,8 @@
 				MX7D_PAD_SD2_DATA1__SD2_DATA1		0x5a
 				MX7D_PAD_SD2_DATA2__SD2_DATA2		0x5a
 				MX7D_PAD_SD2_DATA3__SD2_DATA3		0x5a
+				MX7D_PAD_ECSPI2_MOSI__GPIO4_IO21	0x19 /* WL_REG_ON */
+				MX7D_PAD_ECSPI2_SCLK__GPIO4_IO20	0x19 /* WL_HOST_WAKE */
 			>;
 		};
 
@@ -1186,6 +1191,8 @@
 				MX7D_PAD_SD2_DATA1__SD2_DATA1		0x5b
 				MX7D_PAD_SD2_DATA2__SD2_DATA2		0x5b
 				MX7D_PAD_SD2_DATA3__SD2_DATA3		0x5b
+				MX7D_PAD_ECSPI2_MOSI__GPIO4_IO21	0x19 /* WL_REG_ON */
+				MX7D_PAD_ECSPI2_SCLK__GPIO4_IO20	0x19 /* WL_HOST_WAKE */
 			>;
 		};
 
diff --git a/arch/arm/boot/dts/imx7d.dtsi b/arch/arm/boot/dts/imx7d.dtsi
index ad2fce8..c8cd9b3 100644
--- a/arch/arm/boot/dts/imx7d.dtsi
+++ b/arch/arm/boot/dts/imx7d.dtsi
@@ -303,6 +303,12 @@
 		status = "disabled";
 	};
 
+	pcie_phy: pcie-phy@306d0000 {
+		  compatible = "fsl,imx-pcie-phy";
+		  reg = <0x306d0000 0x10000>;
+		  status = "disabled";
+	};
+
 	epdc: epdc@306f0000 {
 		compatible = "fsl,imx7d-epdc";
 		interrupts = <GIC_SPI 117 IRQ_TYPE_LEVEL_HIGH>;
diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index add632d..4eae0f9 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -90,6 +90,8 @@ dtb-$(CONFIG_ARCH_FSL_IMX8MQ) += fsl-imx8mq-ddr3l-arm2.dtb \
 				 fsl-imx8mq-evk-dp.dtb \
 				 fsl-imx8mq-evk-edp.dtb
 dtb-$(CONFIG_ARCH_FSL_IMX8MM) += fsl-imx8mm-evk.dtb \
+				 imx8mm-evk-usd-m2.dtb \
+				 imx8mm-evk-usd-m2-oob.dtb \
 				 fsl-imx8mm-evk-ak4497.dtb \
 				 fsl-imx8mm-evk-m4.dtb \
 				 fsl-imx8mm-evk-ak5558.dtb \
diff --git a/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2-oob.dts b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2-oob.dts
new file mode 100644
index 00000000..c10b4ac
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2-oob.dts
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2019 NXP
+ */
+
+#include "fsl-imx8mm-evk.dts"
+
+/ {
+  	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		reset-gpios = <&gpio5 11 GPIO_ACTIVE_LOW>;  /* WL REG ON */
+	};
+
+	modem_reset_usd_m2: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 12 GPIO_ACTIVE_LOW>;  /* BT REG ON */
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+};
+
+&reg_usdhc2_vmmc {
+  	regulator-always-on;
+};
+
+&uart1 {
+	/* Onboard 1PJ BT */
+	status = "disabled";
+};
+
+&uart3 {
+  	resets = <&modem_reset_usd_m2>;
+};
+
+&usdhc1 {
+	/* Disable 1PJ */
+	status = "disabled";
+};
+
+
+&pinctrl_usdhc2_gpio {
+	fsl,pins = <
+    		MX8MM_IOMUXC_GPIO1_IO15_GPIO1_IO15	0x1c4
+		MX8MM_IOMUXC_ECSPI2_MOSI_GPIO5_IO11	0x19  /* WL_REG_ON */
+    		MX8MM_IOMUXC_ECSPI2_MISO_GPIO5_IO12	0x19  /* BT_REG_ON */
+    		MX8MM_IOMUXC_GPIO1_IO04_GPIO1_IO4	0x1d0 /* Murata -- should be pull-up on this */
+    		MX8MM_IOMUXC_ECSPI2_SCLK_GPIO5_IO10 	0x1d0 /* Murata WL_HOST_WAKE -- pull-up */
+	>;
+};
+
+&pinctrl_usdhc2 {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x190
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d0
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d0
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d0
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d0
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d0
+	>;
+};
+
+&pinctrl_usdhc2_100mhz {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x194
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d4
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d4
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d4
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d4
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d4
+	>;
+};
+
+&pinctrl_usdhc2_200mhz {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x196
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d6
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d6
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d6
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d6
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d6
+	>;
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	/delete-property/ cd-gpios;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	keep-power-in-suspend;
+	non-removable;
+	cap-power-off-card;
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	  	interrupt-parent = <&gpio5>;
+    		interrupts = <10 IRQ_TYPE_LEVEL_LOW>;
+    		interrupt-names = "host-wake";
+	};
+};
diff --git a/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts
new file mode 100644
index 00000000..c10b4ac
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2019 NXP
+ */
+
+#include "fsl-imx8mm-evk.dts"
+
+/ {
+  	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		reset-gpios = <&gpio5 11 GPIO_ACTIVE_LOW>;  /* WL REG ON */
+	};
+
+	modem_reset_usd_m2: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 12 GPIO_ACTIVE_LOW>;  /* BT REG ON */
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+};
+
+&reg_usdhc2_vmmc {
+  	regulator-always-on;
+};
+
+&uart1 {
+	/* Onboard 1PJ BT */
+	status = "disabled";
+};
+
+&uart3 {
+  	resets = <&modem_reset_usd_m2>;
+};
+
+&usdhc1 {
+	/* Disable 1PJ */
+	status = "disabled";
+};
+
+
+&pinctrl_usdhc2_gpio {
+	fsl,pins = <
+    		MX8MM_IOMUXC_GPIO1_IO15_GPIO1_IO15	0x1c4
+		MX8MM_IOMUXC_ECSPI2_MOSI_GPIO5_IO11	0x19  /* WL_REG_ON */
+    		MX8MM_IOMUXC_ECSPI2_MISO_GPIO5_IO12	0x19  /* BT_REG_ON */
+    		MX8MM_IOMUXC_GPIO1_IO04_GPIO1_IO4	0x1d0 /* Murata -- should be pull-up on this */
+    		MX8MM_IOMUXC_ECSPI2_SCLK_GPIO5_IO10 	0x1d0 /* Murata WL_HOST_WAKE -- pull-up */
+	>;
+};
+
+&pinctrl_usdhc2 {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x190
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d0
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d0
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d0
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d0
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d0
+	>;
+};
+
+&pinctrl_usdhc2_100mhz {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x194
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d4
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d4
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d4
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d4
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d4
+	>;
+};
+
+&pinctrl_usdhc2_200mhz {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x196
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d6
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d6
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d6
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d6
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d6
+	>;
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	/delete-property/ cd-gpios;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	keep-power-in-suspend;
+	non-removable;
+	cap-power-off-card;
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	  	interrupt-parent = <&gpio5>;
+    		interrupts = <10 IRQ_TYPE_LEVEL_LOW>;
+    		interrupt-names = "host-wake";
+	};
+};
