diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 0ec6f5b..85be7c3 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -444,6 +444,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6dl-sabresd.dtb \
 	imx6dl-sabresd-ldo.dtb \
 	imx6dl-sabresd-btwifi.dtb \
+	imx6dl-sabresd-btwifi-m2.dtb \
+	imx6dl-sabresd-btwifi-m2-oob.dtb \
 	imx6dl-sabresd-hdcp.dtb \
 	imx6dl-sabresd-enetirq.dtb \
 	imx6dl-sabresd-pcie.dtb \
@@ -534,6 +536,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6q-sabresd.dtb \
 	imx6q-sabresd-ldo.dtb \
 	imx6q-sabresd-btwifi.dtb \
+	imx6q-sabresd-btwifi-m2.dtb \
+	imx6q-sabresd-btwifi-m2-oob.dtb \
 	imx6q-sabresd-hdcp.dtb \
 	imx6q-sabresd-uart.dtb \
 	imx6q-sabresd-enetirq.dtb \
@@ -570,6 +574,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6qp-sabresd.dtb \
 	imx6qp-sabresd-ldo.dtb \
 	imx6qp-sabresd-btwifi.dtb \
+	imx6qp-sabresd-btwifi-m2.dtb \
+	imx6qp-sabresd-btwifi-m2-oob.dtb \
 	imx6qp-sabresd-hdcp.dtb \
 	imx6qp-sabresd-pcie-ep.dtb \
 	imx6qp-tx6qp-8037.dtb \
@@ -584,6 +590,7 @@ dtb-$(CONFIG_SOC_IMX6SL) += \
 	imx6sl-evk-csi.dtb \
 	imx6sl-evk-uart.dtb \
 	imx6sl-evk-btwifi.dtb \
+	imx6sl-evk-btwifi-m2.dtb \
 	imx6sl-warp.dtb
 dtb-$(CONFIG_SOC_IMX6SLL) += \
 	imx6sll-evk.dtb \
@@ -602,6 +609,8 @@ dtb-$(CONFIG_SOC_IMX6SX) += \
 	imx6sx-sdb-m4.dtb \
 	imx6sx-sdb-mqs.dtb \
 	imx6sx-sdb-btwifi.dtb \
+	imx6sx-sdb-btwifi-m2.dtb \
+	imx6sx-sdb-btwifi-m2-oob.dtb \
 	imx6sx-sdb-pcie-ep.dtb \
 	imx6sx-softing-vining-2000.dtb \
 	imx6sx-udoo-neo-basic.dtb \
@@ -613,6 +622,10 @@ dtb-$(CONFIG_SOC_IMX6UL) += \
 	imx6ul-14x14-evk-emmc.dtb \
 	imx6ul-14x14-evk-btwifi.dtb \
 	imx6ul-14x14-evk-btwifi-oob.dtb \
+	imx6ul-14x14-evk-btwifi-m2.dtb \
+	imx6ul-14x14-evk-btwifi-m2-oob.dtb \
+	imx6ul-14x14-evk-btwifi-m2-uhs.dtb \
+	imx6ul-14x14-evk-btwifi-m2-oob-uhs.dtb \
 	imx6ul-14x14-evk-ecspi-slave.dtb \
 	imx6ul-14x14-evk-ecspi.dtb \
 	imx6ul-14x14-evk-gpmi-weim.dtb \
@@ -639,6 +652,10 @@ dtb-$(CONFIG_SOC_IMX6UL) += \
 	imx6ull-14x14-evk-emmc.dtb \
 	imx6ull-14x14-evk-btwifi.dtb \
 	imx6ull-14x14-evk-btwifi-oob.dtb \
+	imx6ull-14x14-evk-btwifi-m2.dtb \
+	imx6ull-14x14-evk-btwifi-m2-oob.dtb \
+	imx6ull-14x14-evk-btwifi-m2-uhs.dtb \
+	imx6ull-14x14-evk-btwifi-m2-oob-uhs.dtb \
 	imx6ull-14x14-evk-gpmi-weim.dtb \
 	imx6ull-9x9-evk.dtb \
 	imx6ull-9x9-evk-ldo.dtb \
diff --git a/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..602f401
--- /dev/null
+++ b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6dl-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..4742495
--- /dev/null
+++ b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6dl-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..2e8b621
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6q-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..48845c8
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi
@@ -0,0 +1,16 @@
+&iomuxc {
+	imx6qdl-sabresd-murata-v2 {
+		/* add MUXing entry for SD2 4-bit interface and configure control pins */
+		pinctrl_wifi_oob: pinctrl_wifi {
+			fsl,pins = <
+				MX6QDL_PAD_KEY_COL0__GPIO4_IO06		0x0b000	/* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+
+&brcmf{
+	interrupt-parent = <&gpio4>;
+	interrupts = <6 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..7985b59
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6q-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi
new file mode 100644
index 00000000..a74bad0
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/*
+ * NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into SD2
+ * slot using Murata i.MX InterConnect Ver 2.0 Adapter & connecting Bluetooth
+ * UART & control signals via ribbon cable.
+ * This configuration supports both WLAN and Bluetooth.
+ * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected via ribbon cable (J13 connector
+ * on board).
+ * ==> Hardware modification is required. Refer to schematic.
+ *
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration.
+ */
+
+/ {
+	leds {
+		compatible = "gpio-leds";
+		status = "disabled";
+	};
+
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio1 2 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio4 7 GPIO_ACTIVE_LOW>;
+ 	};
+};
+
+&ecspi1 {
+	status = "disabled";
+};
+
+&iomuxc {
+	imx6qdl-sabresd-murata-v2 {
+		pinctrl_btreg: btreggrp {
+			fsl,pins = <
+				MX6QDL_PAD_GPIO_2__GPIO1_IO02 0x1b0b0  /* BT_REG_ON */
+			>;
+		};
+
+		/* add MUXing entry for SD2 4-bit interface and configure control pins */
+		pinctrl_wifi: wifigrp {
+			fsl,pins = <
+				MX6QDL_PAD_SD2_CMD__SD2_CMD		0x17069
+				MX6QDL_PAD_SD2_CLK__SD2_CLK		0x10069
+				MX6QDL_PAD_SD2_DAT0__SD2_DATA0		0x17069
+				MX6QDL_PAD_SD2_DAT1__SD2_DATA1		0x17069
+				MX6QDL_PAD_SD2_DAT2__SD2_DATA2		0x17069
+				MX6QDL_PAD_SD2_DAT3__SD2_DATA3		0x17069
+				MX6QDL_PAD_KEY_ROW0__GPIO4_IO07		0x13069	/* WL_REG_ON */
+				MX6QDL_PAD_KEY_COL0__GPIO4_IO06		0x13069	/* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+
+/* re-defining gpio_leds removes conflict with BT_REG_ON */
+&pinctrl_gpio_leds {
+	fsl,pins = <
+	>;
+};
+
+&uart5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart5_1
+		     &pinctrl_btreg>;
+	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode; */
+	/* pinctrl-0 = <&pinctrl_uart5dte_1>; */
+};
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_wifi>;
+	bus-width = <4>;
+	no-1-8-v;
+	non-removable;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..b08baf8
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6qp-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..1c4a097
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6qp-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts
new file mode 100644
index 00000000..ec6c693
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts
@@ -0,0 +1 @@
+#include "imx6sl-evk-btwifi.dts"
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts
new file mode 100644
index 00000000..0a19a8d
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6sx-sdb-btwifi-m2.dts"
+#include "imx6sx-sdb-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..452bbef
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi
@@ -0,0 +1,18 @@
+&iomuxc {
+	imx6sx-sdb-murata-v1_sdext {
+		pinctrl_wifi_oob: pinctrl_wifi {
+			fsl,pins = <
+				// Murata Module control signals -- not connected (R327 and need to disable ENET1 CLK)
+				MX6SX_PAD_ENET1_RX_CLK__GPIO2_IO_4	0x0b000 /* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+&fec1 {
+	status = "disabled";
+};
+&brcmf{
+	interrupt-parent = <&gpio2>;
+	interrupts = <4 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts
new file mode 100644
index 00000000..3038da0
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6sx-sdb.dts"
+#include "imx6sx-sdb-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi
new file mode 100644
index 00000000..4c7d8d7
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/*
+ * NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into SD3
+ * slot using Murata i.MX InterConnect Ver 1.0 Adapter AND SD Card Extender on
+ * SD2 slot. Bluetooth UART connects via SD3 EMMC/MMC Plus pinout.
+ * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE connect via SD Card Extender.
+ */
+
+/ {
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio4 10 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+//		reset-gpios = <&gpio2 4 GPIO_ACTIVE_LOW>;  /* currently not implemented */
+	};
+};
+
+&iomuxc {
+	imx6sx-sdb-murata-wifibt {
+		pinctrl_bt: btgrp {
+			fsl,pins = <
+				MX6SX_PAD_NAND_DATA06__GPIO4_IO_10	0x13069 /* BT_REG_ON */
+			>;
+		};
+
+		pinctrl_wifi: wifigrp {
+			fsl,pins = <
+				/* Murata Module control signals -- not connected (R327 and need to disable ENET1 CLK) */
+//				MX6SX_PAD_ENET1_RX_CLK__GPIO2_IO_4	0x13069 /* WL_REG_ON */
+			>;
+		};
+	};
+};
+
+&uart5 { /* for bluetooth */
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart5 &pinctrl_bt>;
+	uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode;*/
+	/* pinctrl-0 = <&pinctrl_uart5dte_1>; */
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usdhc2 &pinctrl_wifi>;
+	bus-width = <4>;
+	no-1-8-v;	/* force 3.3V VIO */
+	non-removable;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	cap-power-off-card;
+	status = "okay";
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts
new file mode 100644
index 00000000..0616880
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts
@@ -0,0 +1,2 @@
+#include "imx6ul-14x14-evk-btwifi-m2-uhs.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts
new file mode 100644
index 00000000..5826e96
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6ul-14x14-evk-btwifi-m2.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts
new file mode 100644
index 00000000..4d16874
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts
@@ -0,0 +1,10 @@
+#include "imx6ul-14x14-evk-btwifi-m2.dts"
+
+&usdhc1 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
+	/delete-property/ no-1-8-v;
+};
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts
new file mode 100644
index 00000000..ec58eb2e
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts
@@ -0,0 +1,18 @@
+#include "imx6ul-14x14-evk-btwifi.dts"
+
+&usdhc1_pwrseq {
+	/delete-property/ pinctrl-names;
+	/delete-property/ pinctrl-0;
+	post-power-on-delay-ms = <120>;  /* Murata -- add delay before accessing SDIO bus */
+};
+
+&pinctrl_usdhc1 {
+        fsl,pins = <
+            	MX6UL_PAD_GPIO1_IO05__GPIO1_IO05	0x17059 /* SD1 VSELECT */
+        >;
+};
+
+&usdhc1 {
+	pinctrl-assert-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+};
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk.dtsi b/arch/arm/boot/dts/imx6ul-14x14-evk.dtsi
index 9c8e1c9..ea24dbf 100644
--- a/arch/arm/boot/dts/imx6ul-14x14-evk.dtsi
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk.dtsi
@@ -212,6 +212,18 @@
 	phy-mode = "rmii";
 	phy-handle = <&ethphy0>;
 	status = "okay";
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+			micrel,led-mode = <1>;
+			clocks = <&clks IMX6UL_CLK_ENET_REF>;
+			clock-names = "rmii-ref";
+		};
+	};
 };
 
 &fec2 {
@@ -225,13 +237,6 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		ethphy0: ethernet-phy@2 {
-			reg = <2>;
-			micrel,led-mode = <1>;
-			clocks = <&clks IMX6UL_CLK_ENET_REF>;
-			clock-names = "rmii-ref";
-		};
-
 		ethphy1: ethernet-phy@1 {
 			reg = <1>;
 			micrel,led-mode = <1>;
diff --git a/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..951bce3
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2017 NXP
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+&pinctrl_wifi {
+	fsl,pins = <
+		/* MUXing for WL_HOST_WAKE */
+		MX6UL_PAD_ENET2_RX_ER__GPIO2_IO15  0x0b000 /* input: 100K Pull-up */
+	>;
+};
+
+/*
+ * For WL_HOST_WAKE (OOB_IRQ) to function correctly, we must disable
+ * the secondary ethernet port (FEC2). Hardware re-work is to remove
+ * R1633 and populate R1704 with 0 Ohm resistor.
+ * Refer to Murata Hardware Reference Manual for more details.
+ */
+&fec2 {
+	status = "disabled";
+};
+
+&brcmf {
+	interrupt-parent = <&gpio2>;
+	interrupts = <15 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts
new file mode 100644
index 00000000..8da2a49
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts
@@ -0,0 +1,21 @@
+#include "imx6ull-14x14-evk-btwifi-m2-uhs.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
+
+&fec2 {
+	status="disabled";
+};
+
+&fec1 {
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+			micrel,led-mode = <1>;
+			clocks = <&clks IMX6UL_CLK_ENET_REF>;
+			clock-names = "rmii-ref";
+		};
+	};
+};
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts
new file mode 100644
index 00000000..ff47001
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts
@@ -0,0 +1,5 @@
+#include "imx6ull-14x14-evk-btwifi-m2.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
+
+
+
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts
new file mode 100644
index 00000000..8b4112c
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts
@@ -0,0 +1,10 @@
+#include "imx6ull-14x14-evk-btwifi-m2.dts"
+
+&usdhc1 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
+	/delete-property/ no-1-8-v;
+};
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts
new file mode 100644
index 00000000..a9622fe
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts
@@ -0,0 +1,18 @@
+#include "imx6ull-14x14-evk-btwifi.dts"
+
+&usdhc1_pwrseq {
+	/delete-property/ pinctrl-names;
+	/delete-property/ pinctrl-0;
+	post-power-on-delay-ms = <120>;  /* Murata -- add delay before accessing SDIO bus */
+};
+
+&pinctrl_usdhc1 {
+        fsl,pins = <
+            	MX6UL_PAD_GPIO1_IO05__GPIO1_IO05	0x17059 /* SD1 VSELECT */
+        >;
+};
+
+&usdhc1 {
+	pinctrl-assert-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+};
diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 7b27add..b29c200 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -53,14 +53,14 @@ dtb-$(CONFIG_ARCH_MXC) += imx8mm-evk.dtb imx8mm-evk-rpmsg.dtb imx8mm-evk-rm67191
 			  imx8mm-ddr4-evk-rm67191.dtb imx8mm-evk-revb.dtb imx8mm-ddr4-evk-revb.dtb \
 			  imx8mm-ddr4-evk-revb-rm67191.dtb imx8mm-ddr3l-val.dtb \
 			  imx8mm-evk-pcie-ep.dtb imx8mm-ddr4-evk-pcie-ep.dtb \
-			  imx8mm-evk-usd-wifi.dtb
+			  imx8mm-evk-usd-wifi.dtb imx8mm-evk-usd-m2.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mm-evk-ak4497.dtb imx8mm-evk-ak5558.dtb imx8mm-evk-audio-tdm.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mm-evk-8mic-revE.dtb imx8mm-evk-8mic-swpdm.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mm-ab2.dtb imx8mm-ab2-m4.dtb imx8mm-ddr4-ab2.dtb imx8mm-ddr4-ab2-m4.dtb \
 			  imx8mm-ddr4-ab2-revb.dtb imx8mm-ddr4-ab2-m4-revb.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mn-evk.dtb imx8mn-evk-rm67191.dtb imx8mn-ddr4-evk.dtb imx8mn-ddr4-evk-ak5558.dtb \
 			  imx8mn-ddr4-evk-rm67191.dtb imx8mn-ddr4-evk-rpmsg.dtb imx8mn-ddr4-evk-usd-wifi.dtb \
-			  imx8mn-evk-ak5558.dtb imx8mn-evk-rpmsg.dtb imx8mn-evk-8mic-revE.dtb
+			  imx8mn-evk-ak5558.dtb imx8mn-evk-rpmsg.dtb imx8mn-evk-8mic-revE.dtb imx8mn-evk-usd-m2.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mn-ddr4-evk-root.dtb imx8mn-ddr4-evk-inmate.dtb imx8mn-evk-root.dtb imx8mn-evk-inmate.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mn-ab2.dtb imx8mn-ddr4-ab2.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mp-evk.dtb imx8mp-evk-root.dtb imx8mp-evk-inmate.dtb imx8mp-evk-rpmsg.dtb \
diff --git a/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts
new file mode 100644
index 00000000..f84c0d9
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2019 NXP
+ */
+
+/dts-v1/;
+
+#include "imx8mm-evk.dts"
+
+/ {
+  	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		reset-gpios = <&gpio5 11 GPIO_ACTIVE_LOW>;  /* WL REG ON */
+	};
+
+	modem_reset_usd_m2: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 12 GPIO_ACTIVE_LOW>;  /* BT REG ON */
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+};
+
+&reg_usdhc2_vmmc {
+  	regulator-always-on;
+};
+
+&uart1 {
+	/* Onboard 1PJ BT */
+	status = "disabled";
+};
+
+&uart3 {
+  	resets = <&modem_reset_usd_m2>;
+};
+
+&usdhc1 {
+	/* Disable 1PJ */
+	status = "disabled";
+};
+
+
+&pinctrl_usdhc2_gpio {
+	fsl,pins = <
+    		MX8MM_IOMUXC_GPIO1_IO15_GPIO1_IO15	0x1c4
+		MX8MM_IOMUXC_ECSPI2_MOSI_GPIO5_IO11	0x19  /* WL_REG_ON */
+    		MX8MM_IOMUXC_ECSPI2_MISO_GPIO5_IO12	0x19  /* BT_REG_ON */
+    		MX8MM_IOMUXC_GPIO1_IO04_GPIO1_IO4	0x1d0 /* Murata -- should be pull-up on this */
+    		MX8MM_IOMUXC_ECSPI2_SCLK_GPIO5_IO10 	0x1d0 /* Murata WL_HOST_WAKE -- pull-up */
+	>;
+};
+
+&pinctrl_usdhc2 {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x190
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d0
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d0
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d0
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d0
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d0
+	>;
+};
+
+&pinctrl_usdhc2_100mhz {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x194
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d4
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d4
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d4
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d4
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d4
+	>;
+};
+
+&pinctrl_usdhc2_200mhz {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x196
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d6
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d6
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d6
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d6
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d6
+	>;
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	/delete-property/ cd-gpios;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	keep-power-in-suspend;
+	non-removable;
+	cap-power-off-card;
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	  	interrupt-parent = <&gpio5>;
+    		interrupts = <10 IRQ_TYPE_LEVEL_LOW>;
+    		interrupt-names = "host-wake";
+	};
+};
diff --git a/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2.dts b/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2.dts
new file mode 100644
index 00000000..eb692b6
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2.dts
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2019 NXP
+ */
+
+/dts-v1/;
+
+#include "imx8mn-evk.dts"
+
+/ {
+  	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		reset-gpios = <&gpio5 11 GPIO_ACTIVE_LOW>;  /* WL REG ON */
+	};
+
+
+	modem_reset_usd_m2: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 12 GPIO_ACTIVE_LOW>;  /* BT REG ON */
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+};
+
+&reg_usdhc2_vmmc {
+  	regulator-always-on;
+};
+
+&uart1 {
+	/* Onboard 1MW BT */
+	status = "disabled";
+};
+
+&uart3 {
+  	resets = <&modem_reset_usd_m2>;
+};
+
+&ecspi2 {
+	status = "disabled";
+};
+
+&usdhc1 {
+	/* Disable 1MW */
+	status = "disabled";
+};
+
+
+&pinctrl_usdhc2_gpio {
+	fsl,pins = <
+    		MX8MN_IOMUXC_GPIO1_IO15_GPIO1_IO15	0x1c4
+        	MX8MN_IOMUXC_ECSPI2_MOSI_GPIO5_IO11	0x19  /* WL_REG_ON */
+    		MX8MN_IOMUXC_ECSPI2_MISO_GPIO5_IO12	0x19  /* BT_REG_ON */
+    		MX8MN_IOMUXC_GPIO1_IO04_GPIO1_IO4	0x1d0 /* Murata -- should be pull-up on this */
+    		MX8MN_IOMUXC_ECSPI2_SCLK_GPIO5_IO10 	0x1d0 /* Murata WL_HOST_WAKE -- pull-up */
+	>;
+};
+
+&pinctrl_usdhc2 {
+	fsl,pins = <
+		MX8MN_IOMUXC_SD2_CLK_USDHC2_CLK		0x190
+		MX8MN_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d0
+		MX8MN_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d0
+		MX8MN_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d0
+		MX8MN_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d0
+		MX8MN_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d0
+	>;
+};
+
+&pinctrl_usdhc2_100mhz {
+	fsl,pins = <
+		MX8MN_IOMUXC_SD2_CLK_USDHC2_CLK		0x194
+		MX8MN_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d4
+		MX8MN_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d4
+		MX8MN_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d4
+		MX8MN_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d4
+		MX8MN_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d4
+	>;
+};
+
+&pinctrl_usdhc2_200mhz {
+	fsl,pins = <
+		MX8MN_IOMUXC_SD2_CLK_USDHC2_CLK		0x196
+		MX8MN_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d6
+		MX8MN_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d6
+		MX8MN_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d6
+		MX8MN_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d6
+		MX8MN_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d6
+	>;
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	/delete-property/ cd-gpios;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	keep-power-in-suspend;
+	non-removable;
+	cap-power-off-card;
+	brcmf2: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	  	interrupt-parent = <&gpio5>;
+    		interrupts = <10 IRQ_TYPE_LEVEL_LOW>;
+    		interrupt-names = "host-wake";
+	};
+};
