diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
index 4691a43..7a2b5a1 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
@@ -4395,6 +4395,8 @@ static void brcmf_sdio_firmware_callback(struct device *dev, int err,
 					  CY_43455_MESBUSYCTRL, &err);
 			break;
 		case SDIO_DEVICE_ID_BROADCOM_4339:
+			brcmf_dbg(INFO, "set F2 watermark to 0x%x*4 bytes for 4339\n",
+					   CY_4339_F2_WATERMARK);
 			brcmf_sdiod_writeb(sdiod,
 					   SBSDIO_WATERMARK,
 					   CY_4339_F2_WATERMARK, &err);
@@ -4703,4 +4705,3 @@ int brcmf_sdio_sleep(struct brcmf_sdio *bus, bool sleep)
 
 	return ret;
 }
-
