diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
index cc8a7ec..1205b54 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
@@ -2036,7 +2036,7 @@ static int brcmf_fws_borrow_credit(struct brcmf_fws_info *fws,
 
 		if (!borrow_all)
 			borrow_limit =
-			  fws->fifo_credit[lender_ac] / BRCMF_BORROW_RATIO;
+			  fws->init_fifo_credit[lender_ac] / BRCMF_BORROW_RATIO;
 		else
 			borrow_limit = 0;
 
