From 1c341dba65b9ba2585a74a9c47799f80d29e7c4c Mon Sep 17 00:00:00 2001
From: Romit Chatterjee <romit.chatterjee@murata.com>
Date: Wed, 25 Feb 2026 08:17:31 -0600
Subject: [PATCH] wlu: fix inttypes.h include guard for Linux build

Upstream-Status: Inappropriate [proprietary vendor code, no upstream]

---
 wlu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/wl/src/wlu.c b/wl/src/wlu.c
index 01b28d1..f36161c 100644
--- a/wlu.c
+++ b/wlu.c
@@ -123,7 +123,7 @@
 #include <sys/time.h>
 #endif   /* defined(WLBSSLOAD_REPORT) && defined(linux) */
 
-#ifdef LINUX
+#if defined(linux)
 #include <inttypes.h>
 #endif // endif
 #include <miniopt.h>
-- 
2.53.0

