diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 3939c91..2a1423c 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -60,7 +60,8 @@ dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-lx2160a-rdb.dtb
 dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-lx2162a-qds.dtb
 
 dtb-$(CONFIG_ARCH_MXC) += imx8mm-beacon-kit.dtb
-dtb-$(CONFIG_ARCH_MXC) += imx8mm-evk.dtb imx8mm-evk-rpmsg.dtb imx8mm-evk-rm67191.dtb \
+dtb-$(CONFIG_ARCH_MXC) += imx8mm-evk.dtb imx8mm-evk-usd-m2.dtb imx8mm-evk-usd-m2-oob.dtb \
+			  imx8mm-evk-rpmsg.dtb imx8mm-evk-rm67191.dtb \
 			  imx8mm-evk-root.dtb imx8mm-evk-inmate.dtb imx8mm-evk-revb-qca-wifi.dtb \
 			  imx8mm-evk-ecspi-slave.dtb \
 			  imx8mm-evk-pcie-ep.dtb \
@@ -79,7 +80,8 @@ dtb-$(CONFIG_ARCH_MXC) += imx8mm-ddr4-evk.dtb imx8mm-ddr4-evk-rm67191.dtb imx8mm
 			  imx8mm-ddr4-evk-revb-rm67199.dtb \
 			  imx8mm-ddr4-evk-rm67199.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mm-var-som-symphony.dtb
-dtb-$(CONFIG_ARCH_MXC) += imx8mn-evk.dtb imx8mn-evk-rm67191.dtb imx8mn-evk-rpmsg.dtb
+dtb-$(CONFIG_ARCH_MXC) += imx8mn-evk.dtb imx8mn-evk-usd-m2.dtb imx8mn-evk-usd-m2-oob.dtb \
+			  imx8mn-evk-rm67191.dtb imx8mn-evk-rpmsg.dtb \
 dtb-$(CONFIG_ARCH_MXC) += imx8mm-ab2.dtb imx8mm-ab2-m4.dtb imx8mm-ddr4-ab2.dtb imx8mm-ddr4-ab2-m4.dtb \
 			  imx8mm-ddr4-ab2-revb.dtb imx8mm-ddr4-ab2-m4-revb.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx8mn-evk.dtb imx8mn-evk-rm67191.dtb imx8mn-evk-rpmsg.dtb imx8mn-evk-ak5558.dtb \
diff --git a/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2-oob.dts b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2-oob.dts
new file mode 100644
index 0000000..ac85f9f
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2-oob.dts
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2019 NXP
+ */
+
+#include "imx8mm-evk-usd-m2.dts"
+
+
+&pinctrl_usdhc2_gpio {
+	fsl,pins = <
+    		MX8MM_IOMUXC_GPIO1_IO15_GPIO1_IO15	0x1c4
+		MX8MM_IOMUXC_ECSPI2_MOSI_GPIO5_IO11	0x19  /* WL_REG_ON */
+    		MX8MM_IOMUXC_ECSPI2_MISO_GPIO5_IO12	0x19  /* BT_REG_ON */
+    		MX8MM_IOMUXC_GPIO1_IO04_GPIO1_IO4	0x1d0 /* Murata -- should be pull-up on this */
+    		MX8MM_IOMUXC_ECSPI2_SCLK_GPIO5_IO10 	0x1d0 /* Murata WL_HOST_WAKE -- pull-up */
+	>;
+};
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	/delete-property/ cd-gpios;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	keep-power-in-suspend;
+	non-removable;
+	cap-power-off-card;
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	  	interrupt-parent = <&gpio5>;
+    		interrupts = <10 IRQ_TYPE_LEVEL_LOW>;
+    		interrupt-names = "host-wake";
+	};
+};
diff --git a/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts
new file mode 100644
index 0000000..3509e02
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mm-evk-usd-m2.dts
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2019 NXP
+ */
+
+/dts-v1/;
+
+#include "imx8mm-evk.dts"
+
+/ {
+  	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		reset-gpios = <&gpio5 11 GPIO_ACTIVE_LOW>;  /* WL REG ON */
+	};
+
+	modem_reset_usd_m2: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 12 GPIO_ACTIVE_LOW>;  /* BT REG ON */
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+};
+
+&reg_usdhc2_vmmc {
+  	regulator-always-on;
+};
+
+&uart1 {
+	/* Onboard AW-CM358SM BT */
+	status = "disabled";
+};
+
+&uart3 {
+  	resets = <&modem_reset_usd_m2>;
+};
+
+&ecspi2 {
+	status = "disabled";
+};
+
+&usdhc1 {
+	/* Disable AW-CM358SM */
+	status = "disabled";
+};
+
+
+&pinctrl_usdhc2_gpio {
+	fsl,pins = <
+    		MX8MM_IOMUXC_GPIO1_IO15_GPIO1_IO15	0x1c4
+		MX8MM_IOMUXC_ECSPI2_MOSI_GPIO5_IO11	0x19  /* WL_REG_ON */
+    		MX8MM_IOMUXC_ECSPI2_MISO_GPIO5_IO12	0x19  /* BT_REG_ON */
+    		MX8MM_IOMUXC_GPIO1_IO04_GPIO1_IO4	0x1d0 /* Murata -- should be pull-up on this */
+	>;
+};
+
+&pinctrl_usdhc2 {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x190
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d0
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d0
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d0
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d0
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d0
+	>;
+};
+
+&pinctrl_usdhc2_100mhz {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x194
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d4
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d4
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d4
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d4
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d4
+	>;
+};
+
+&pinctrl_usdhc2_200mhz {
+	fsl,pins = <
+		MX8MM_IOMUXC_SD2_CLK_USDHC2_CLK		0x196
+		MX8MM_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d6
+		MX8MM_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d6
+		MX8MM_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d6
+		MX8MM_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d6
+		MX8MM_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d6
+	>;
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	/delete-property/ cd-gpios;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	keep-power-in-suspend;
+	non-removable;
+	cap-power-off-card;
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2-oob.dts b/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2-oob.dts
new file mode 100644
index 0000000..d1826e7
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2-oob.dts
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2019 NXP
+ */
+
+#include "imx8mn-evk-usd-m2.dts"
+
+&pinctrl_usdhc2_gpio {
+	fsl,pins = <
+    		MX8MN_IOMUXC_GPIO1_IO15_GPIO1_IO15	0x1c4
+		MX8MN_IOMUXC_ECSPI2_MOSI_GPIO5_IO11	0x19  /* WL_REG_ON */
+    		MX8MN_IOMUXC_ECSPI2_MISO_GPIO5_IO12	0x19  /* BT_REG_ON */
+    		MX8MN_IOMUXC_GPIO1_IO04_GPIO1_IO4	0x1d0 /* Murata -- should be pull-up on this */
+    		MX8MN_IOMUXC_ECSPI2_SCLK_GPIO5_IO10 	0x1d0 /* Murata WL_HOST_WAKE -- pull-up */
+	>;
+};
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	/delete-property/ cd-gpios;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	keep-power-in-suspend;
+	non-removable;
+	cap-power-off-card;
+	brcmf2: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	  	interrupt-parent = <&gpio5>;
+    		interrupts = <10 IRQ_TYPE_LEVEL_LOW>;
+    		interrupt-names = "host-wake";
+	};
+};
diff --git a/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2.dts b/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2.dts
new file mode 100644
index 0000000..c37c29c
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mn-evk-usd-m2.dts
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2019 NXP
+ */
+
+/dts-v1/;
+
+#include "imx8mn-evk.dts"
+
+/ {
+  	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		reset-gpios = <&gpio5 11 GPIO_ACTIVE_LOW>;  /* WL REG ON */
+	};
+
+	modem_reset_usd_m2: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 12 GPIO_ACTIVE_LOW>;  /* BT REG ON */
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+};
+
+&reg_usdhc2_vmmc {
+  	regulator-always-on;
+};
+
+&uart1 {
+	/* Onboard AW-CM358SM BT */
+	status = "disabled";
+};
+
+&uart3 {
+  	resets = <&modem_reset_usd_m2>;
+};
+
+&ecspi2 {
+	status = "disabled";
+};
+
+&usdhc1 {
+	/* Disable AW-CM358SM */
+	status = "disabled";
+};
+
+
+&pinctrl_usdhc2_gpio {
+	fsl,pins = <
+    		MX8MN_IOMUXC_GPIO1_IO15_GPIO1_IO15	0x1c4
+		MX8MN_IOMUXC_ECSPI2_MOSI_GPIO5_IO11	0x19  /* WL_REG_ON */
+    		MX8MN_IOMUXC_ECSPI2_MISO_GPIO5_IO12	0x19  /* BT_REG_ON */
+    		MX8MN_IOMUXC_GPIO1_IO04_GPIO1_IO4	0x1d0 /* Murata -- should be pull-up on this */
+	>;
+};
+
+&pinctrl_usdhc2 {
+	fsl,pins = <
+		MX8MN_IOMUXC_SD2_CLK_USDHC2_CLK		0x190
+		MX8MN_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d0
+		MX8MN_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d0
+		MX8MN_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d0
+		MX8MN_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d0
+		MX8MN_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d0
+	>;
+};
+
+&pinctrl_usdhc2_100mhz {
+	fsl,pins = <
+		MX8MN_IOMUXC_SD2_CLK_USDHC2_CLK		0x194
+		MX8MN_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d4
+		MX8MN_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d4
+		MX8MN_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d4
+		MX8MN_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d4
+		MX8MN_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d4
+	>;
+};
+
+&pinctrl_usdhc2_200mhz {
+	fsl,pins = <
+		MX8MN_IOMUXC_SD2_CLK_USDHC2_CLK		0x196
+		MX8MN_IOMUXC_SD2_CMD_USDHC2_CMD		0x1d6
+		MX8MN_IOMUXC_SD2_DATA0_USDHC2_DATA0	0x1d6
+		MX8MN_IOMUXC_SD2_DATA1_USDHC2_DATA1	0x1d6
+		MX8MN_IOMUXC_SD2_DATA2_USDHC2_DATA2	0x1d6
+		MX8MN_IOMUXC_SD2_DATA3_USDHC2_DATA3	0x1d6
+	>;
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	/delete-property/ cd-gpios;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	keep-power-in-suspend;
+	non-removable;
+	cap-power-off-card;
+	brcmf2: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
