From 1f38b7a4860b558f350016eea2c3aade109839a6 Mon Sep 17 00:00:00 2001
From: Romit Chatterjee <romit@alumnux.com>
Date: Mon, 31 Aug 2026 10:24:39 +0530
Subject: [PATCH] cfg80211: disable signed regulatory database by default

Upstream-Status: Inappropriate [Murata platform configuration]

---
 net/wireless/Kconfig | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/wireless/Kconfig b/net/wireless/Kconfig
index 2d67b5f2010e..605fd7d764fe 100644
--- a/net/wireless/Kconfig
+++ b/net/wireless/Kconfig
@@ -88,7 +88,7 @@ config CFG80211_CERTIFICATION_ONUS
 
 config CFG80211_REQUIRE_SIGNED_REGDB
 	bool "require regdb signature" if CFG80211_CERTIFICATION_ONUS
-	default y
+	default n
 	select SYSTEM_DATA_VERIFICATION
 	help
 	  Require that in addition to the "regulatory.db" file a
@@ -98,7 +98,7 @@ config CFG80211_REQUIRE_SIGNED_REGDB
 
 config CFG80211_USE_KERNEL_REGDB_KEYS
 	bool "allow regdb keys shipped with the kernel" if CFG80211_CERTIFICATION_ONUS
-	default y
+	default n
 	depends on CFG80211_REQUIRE_SIGNED_REGDB
 	help
 	  Allow the regulatory database to be signed by one of the keys for
