diff --git a/drivers/tty/serial/imx.c b/drivers/tty/serial/imx.c
index e9c214b0ee27..0204c973b71f 100644
--- a/drivers/tty/serial/imx.c
+++ b/drivers/tty/serial/imx.c
@@ -1492,10 +1492,10 @@ static int imx_uart_startup(struct uart_port *port)
 
 	imx_uart_writel(sport, ucr4 & ~UCR4_DREN, UCR4);
 
-	/* Can we enable the DMA support? */
-	if (!uart_console(port) && imx_uart_dma_init(sport) == 0)
-		dma_is_inited = 1;
-
+/*  // Disable DMA support as the Kernel crashes when we make HCI_UART as module
+ 	if (!uart_console(port) && imx_uart_dma_init(sport) == 0)
+ 		dma_is_inited = 1;
+*/
 	spin_lock_irqsave(&sport->port.lock, flags);
 
 	/* Reset fifo's and state machines */
