diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/bcmsdh.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/bcmsdh.c
index 426bdd5..d89a6aa 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/bcmsdh.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/bcmsdh.c
@@ -1148,9 +1148,6 @@ static int brcmf_ops_sdio_suspend(struct device *dev)
 
 	brcmf_dbg(SDIO, "Enter: F%d\n", func->num);
 
-	if (func->num != 1)
-		return 0;
-
 	while (retry &&
 	       config->pm_state == BRCMF_CFG80211_PM_STATE_SUSPENDING) {
 		usleep_range(10000, 20000);
@@ -1159,6 +1156,9 @@ static int brcmf_ops_sdio_suspend(struct device *dev)
 	if (!retry && config->pm_state == BRCMF_CFG80211_PM_STATE_SUSPENDING)
 		brcmf_err("timed out wait for cfg80211 suspended\n");
 
+	if (func->num != 1)
+		return 0;
+
 	sdiodev = bus_if->bus_priv.sdio;
 
 	brcmf_sdiod_freezer_on(sdiodev);
@@ -1226,4 +1226,3 @@ void brcmf_sdio_exit(void)
 
 	sdio_unregister_driver(&brcmf_sdmmc_driver);
 }
-
