diff --git a/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi b/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
index f2a6835..b640a33 100644
--- a/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
+++ b/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
@@ -45,9 +45,12 @@
 &usdhc1 {
 	#address-cells = <1>;
 	#size-cells = <0>;
-	pinctrl-names = "default";
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
 	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
-	no-1-8-v;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+//	no-1-8-v; /* Murata - comment out for UHS mode */
+	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
 	non-removable;
 	pm-ignore-notify;
 	mmc-pwrseq = <&usdhc1_pwrseq>;
