diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 52418b5..57dc25b 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -549,6 +549,8 @@ dtb-$(CONFIG_SOC_IMX6UL) += \
 	imx6ul-14x14-evk-btwifi-oob.dtb \
 	imx6ul-14x14-evk-btwifi-m2.dtb \
 	imx6ul-14x14-evk-btwifi-m2-oob.dtb \
+	imx6ul-14x14-evk-btwifi-m2-50MHz.dtb \
+	imx6ul-14x14-evk-btwifi-m2-oob-50MHz.dtb \
 	imx6ul-14x14-evk-csi.dtb \
 	imx6ul-14x14-evk-emmc.dtb \
 	imx6ul-14x14-evk-ecspi.dtb \
@@ -602,6 +604,8 @@ dtb-$(CONFIG_SOC_IMX6ULL) += \
 	imx6ull-14x14-evk-btwifi-oob.dtb \
 	imx6ull-14x14-evk-btwifi-m2.dtb \
 	imx6ull-14x14-evk-btwifi-m2-oob.dtb \
+	imx6ull-14x14-evk-btwifi-m2-50MHz.dtb \
+	imx6ull-14x14-evk-btwifi-m2-oob-50MHz.dtb \
 	imx6ull-14x14-evk-emmc.dtb \
 	imx6ull-14x14-evk-gpmi-weim.dtb \
 	imx6ull-14x14-evk-usb-certi.dtb \
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-50MHz.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-50MHz.dts
new file mode 100644
index 00000000..f5a9111
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-50MHz.dts
@@ -0,0 +1,10 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include "imx6ul-14x14-evk.dts"
+#include "imx6ul-evk-btwifi-50MHz.dtsi"
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-50MHz.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-50MHz.dts
new file mode 100644
index 00000000..ce439fa
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-50MHz.dts
@@ -0,0 +1 @@
+#include "imx6ul-14x14-evk-btwifi-50MHz.dts"
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-50MHz.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-50MHz.dts
new file mode 100644
index 00000000..9d34d94
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-50MHz.dts
@@ -0,0 +1,2 @@
+#include "imx6ul-14x14-evk-btwifi-m2-50MHz.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6ul-evk-btwifi-50MHz.dtsi b/arch/arm/boot/dts/imx6ul-evk-btwifi-50MHz.dtsi
new file mode 100644
index 00000000..de3a9b9
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-evk-btwifi-50MHz.dtsi
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/*
+ * NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into Slot
+ * SD1 and using Murata i.MX InterConnect Ver 2.0 Adapter. Bluetooth UART &
+ * control signals are connected via ribbon cable (J1701 connector).
+ */
+
+/ {
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio_spi 4 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 1 GPIO_ACTIVE_LOW>;
+	};
+};
+
+&iomuxc {
+	pinctrl_wifi: wifigrp {
+		fsl,pins = <
+			MX6UL_PAD_SNVS_TAMPER1__GPIO5_IO01 0x03029  /* WL_REG_ON */
+		>;
+	};
+};
+
+&reg_sd1_vmmc {
+	regulator-always-on;
+};
+
+&uart2 {
+	resets = <&modem_reset>;  /* Murata - add so BT_REG_ON is driven high during kernel boot */
+};
+
+&usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+//	no-1-8-v; /* Murata - comment out for UHS mode */
+	max-frequency = <50000000>; /* Murata - Limit max-frequency to 50MHz for non UHS mode */
+	non-removable;
+	pm-ignore-notify;
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+	cap-power-off-card;
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
+
+&gpio_spi {
+	/* Murata: modify default setting so that BT_nPWD/BT_REG_ON
+	 * is low (0V) during kernel boot.
+	 */
+	registers-default = /bits/ 8 <0x47>;
+};
diff --git a/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi b/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
index f2a6835..b640a33 100644
--- a/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
+++ b/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
@@ -45,9 +45,12 @@
 &usdhc1 {
 	#address-cells = <1>;
 	#size-cells = <0>;
-	pinctrl-names = "default";
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
 	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
-	no-1-8-v;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+//	no-1-8-v; /* Murata - comment out for UHS mode */
+	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
 	non-removable;
 	pm-ignore-notify;
 	mmc-pwrseq = <&usdhc1_pwrseq>;
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-50MHz.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-50MHz.dts
new file mode 100644
index 00000000..3a56921
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-50MHz.dts
@@ -0,0 +1,10 @@
+/*
+ * Copyright (C) 2016 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include "imx6ull-14x14-evk.dts"
+#include "imx6ul-evk-btwifi-50MHz.dtsi"
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-50MHz.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-50MHz.dts
new file mode 100644
index 00000000..65c5950
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-50MHz.dts
@@ -0,0 +1 @@
+#include "imx6ull-14x14-evk-btwifi-50MHz.dts"
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-50MHz.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-50MHz.dts
new file mode 100644
index 00000000..ddb2330
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-50MHz.dts
@@ -0,0 +1,2 @@
+#include "imx6ull-14x14-evk-btwifi-m2-50MHz.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
