diff --git a/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi b/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
index b640a33..de3a9b9 100644
--- a/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
+++ b/arch/arm/boot/dts/imx6ul-evk-btwifi.dtsi
@@ -50,7 +50,7 @@
 	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
 	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
 //	no-1-8-v; /* Murata - comment out for UHS mode */
-	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
+	max-frequency = <50000000>; /* Murata - Limit max-frequency to 50MHz for non UHS mode */
 	non-removable;
 	pm-ignore-notify;
 	mmc-pwrseq = <&usdhc1_pwrseq>;
