diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
index 1492cc5..bb6b632 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/sdio.c
@@ -3537,6 +3537,7 @@ static void brcmf_sdio_sr_init(struct brcmf_sdio *bus)
 
 	if (bus->ci->chip == CY_CC_43012_CHIP_ID ||
 	    bus->ci->chip == BRCM_CC_4339_CHIP_ID ||
+           bus->ci->chip == BRCM_CC_4354_CHIP_ID ||
 	    bus->ci->chip == BRCM_CC_4345_CHIP_ID) {
 		cardcap = SDIO_CCCR_BRCM_CARDCAP_CMD_NODEC;
 	} else {
