diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
index 49806d5..a5503d6 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
@@ -2041,7 +2041,7 @@ static int brcmf_fws_borrow_credit(struct brcmf_fws_info *fws,
 	for (lender_ac = 0; lender_ac <= highest_lender_ac; lender_ac++) {
 		if (!borrow_all)
 			borrow_limit =
-			  fws->fifo_credit[lender_ac] / BRCMF_BORROW_RATIO;
+			  fws->init_fifo_credit[lender_ac] / BRCMF_BORROW_RATIO;
 		else
 			borrow_limit = 0;
 
