diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
index 63ab4e3..bb7fd15 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
@@ -2301,7 +2301,7 @@ static void brcmf_fws_dequeue_worker(struct work_struct *worker)
 		    !fws->bus_flow_blocked) {
 			while (brcmf_fws_borrow_credit(fws,
 						       fifo - 1, fifo,
-						       false) == 0) {
+						       true) == 0) {
 				skb = brcmf_fws_deq(fws, fifo);
 				if (!skb) {
 					brcmf_fws_return_credits(fws, fifo, 1);
