diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 742a8f5..9e1cdaf 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -276,6 +276,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6dl-sabrelite.dtb \
 	imx6dl-sabresd.dtb \
 	imx6dl-sabresd-btwifi.dtb \
+	imx6dl-sabresd-btwifi-m2.dtb \
+	imx6dl-sabresd-btwifi-m2-oob.dtb \
 	imx6dl-sabresd-hdcp.dtb \
 	imx6dl-sabresd-ldo.dtb \
 	imx6dl-tx6dl-comtft.dtb \
@@ -314,12 +316,16 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6qp-sabreauto-gpmi-weim.dtb \
 	imx6qp-sabresd.dtb \
 	imx6qp-sabresd-btwifi.dtb \
+	imx6qp-sabresd-btwifi-m2.dtb \
+	imx6qp-sabresd-btwifi-m2-oob.dtb \
 	imx6qp-sabresd-hdcp.dtb \
 	imx6qp-sabresd-ldo.dtb \
 	imx6qp-sabresd-ldo-pcie-cert.dtb \
 	imx6q-sabrelite.dtb \
 	imx6q-sabresd.dtb \
 	imx6q-sabresd-btwifi.dtb \
+	imx6q-sabresd-btwifi-m2.dtb \
+	imx6q-sabresd-btwifi-m2-oob.dtb \
 	imx6q-sabresd-hdcp.dtb \
 	imx6q-sabresd-ldo.dtb \
 	imx6q-sabresd-enetirq.dtb \
@@ -348,6 +354,7 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 dtb-$(CONFIG_SOC_IMX6SL) += \
 	imx6sl-evk.dtb \
 	imx6sl-evk-btwifi.dtb \
+	imx6sl-evk-btwifi-m2.dtb \
 	imx6sl-evk-ldo.dtb \
 	imx6sl-evk-csi.dtb \
 	imx6sl-evk-uart.dtb \
@@ -360,6 +367,8 @@ dtb-$(CONFIG_SOC_IMX6SX) += \
 	imx6sx-sdb-reva-ldo.dtb \
 	imx6sx-sdb.dtb	\
 	imx6sx-sdb-btwifi.dtb \
+	imx6sx-sdb-btwifi-m2.dtb \
+	imx6sx-sdb-btwifi-m2-oob.dtb \
 	imx6sx-sdb-emmc.dtb \
 	imx6sx-sdb-lcdif1.dtb \
 	imx6sx-sdb-ldo.dtb	\
@@ -388,6 +397,10 @@ dtb-$(CONFIG_SOC_IMX6UL) += \
 	imx6ul-14x14-ddr3-arm2-wm8958.dtb \
 	imx6ul-14x14-evk.dtb \
 	imx6ul-14x14-evk-btwifi.dtb \
+	imx6ul-14x14-evk-btwifi-m2.dtb \
+	imx6ul-14x14-evk-btwifi-m2-oob.dtb \
+	imx6ul-14x14-evk-btwifi-m2-uhs.dtb \
+	imx6ul-14x14-evk-btwifi-m2-oob-uhs.dtb \
 	imx6ul-14x14-evk-csi.dtb \
 	imx6ul-14x14-evk-emmc.dtb \
 	imx6ul-14x14-evk-gpmi-weim.dtb \
@@ -416,6 +429,10 @@ dtb-$(CONFIG_SOC_IMX6ULL) += \
 	imx6ull-14x14-ddr3-arm2-wm8958.dtb \
 	imx6ull-14x14-evk.dtb \
 	imx6ull-14x14-evk-btwifi.dtb \
+	imx6ull-14x14-evk-btwifi-m2.dtb \
+	imx6ull-14x14-evk-btwifi-m2-oob.dtb \
+	imx6ull-14x14-evk-btwifi-m2-uhs.dtb \
+	imx6ull-14x14-evk-btwifi-m2-oob-uhs.dtb \
 	imx6ull-14x14-evk-emmc.dtb \
 	imx6ull-14x14-evk-gpmi-weim.dtb \
 	imx6ull-14x14-evk-usb-certi.dtb \
diff --git a/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..602f401
--- /dev/null
+++ b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6dl-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..4742495
--- /dev/null
+++ b/arch/arm/boot/dts/imx6dl-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6dl-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..2e8b621
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6q-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..48845c8
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2-oob.dtsi
@@ -0,0 +1,16 @@
+&iomuxc {
+	imx6qdl-sabresd-murata-v2 {
+		/* add MUXing entry for SD2 4-bit interface and configure control pins */
+		pinctrl_wifi_oob: pinctrl_wifi {
+			fsl,pins = <
+				MX6QDL_PAD_KEY_COL0__GPIO4_IO06		0x0b000	/* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+
+&brcmf{
+	interrupt-parent = <&gpio4>;
+	interrupts = <6 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..7985b59
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6q-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi
new file mode 100644
index 00000000..08d9098
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi-m2.dtsi
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/*
+ * NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into SD2
+ * slot using Murata i.MX InterConnect Ver 2.0 Adapter & connecting Bluetooth
+ * UART & control signals via ribbon cable.
+ * This configuration supports both WLAN and Bluetooth.
+ * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected via ribbon cable (J13 connector
+ * on board).
+ * ==> Hardware modification is required. Refer to schematic.
+ *
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration.
+ */
+
+/ {
+	leds {
+		compatible = "gpio-leds";
+		status = "disabled";
+	};
+
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio1 2 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio4 7 GPIO_ACTIVE_LOW>;
+ 	};
+};
+
+&ecspi1 {
+	status = "disabled";
+};
+
+&iomuxc {
+	imx6qdl-sabresd-murata-v2 {
+		pinctrl_btreg: btreggrp {
+			fsl,pins = <
+				MX6QDL_PAD_GPIO_2__GPIO1_IO02 0x1b0b0  /* BT_REG_ON */
+			>;
+		};
+
+		/* add MUXing entry for SD2 4-bit interface and configure control pins */
+		pinctrl_wifi: wifigrp {
+			fsl,pins = <
+				MX6QDL_PAD_SD2_CMD__SD2_CMD		0x17059
+				MX6QDL_PAD_SD2_CLK__SD2_CLK		0x10059
+				MX6QDL_PAD_SD2_DAT0__SD2_DATA0		0x17059
+				MX6QDL_PAD_SD2_DAT1__SD2_DATA1		0x17059
+				MX6QDL_PAD_SD2_DAT2__SD2_DATA2		0x17059
+				MX6QDL_PAD_SD2_DAT3__SD2_DATA3		0x17059
+				MX6QDL_PAD_KEY_ROW0__GPIO4_IO07		0x13069	/* WL_REG_ON */
+				MX6QDL_PAD_KEY_COL0__GPIO4_IO06		0x13069	/* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+
+/* re-defining gpio_leds removes conflict with BT_REG_ON */
+&pinctrl_gpio_leds {
+	fsl,pins = <
+	>;
+};
+
+&uart5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart5_1
+		     &pinctrl_btreg>;
+	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode; */
+	/* pinctrl-0 = <&pinctrl_uart5dte_1>; */
+};
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_wifi>;
+	bus-width = <4>;
+	no-1-8-v;
+	non-removable;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm/boot/dts/imx6qdl-sabresd-btwifi.dtsi b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi.dtsi
index 5e697a0..0bfa9e9 100644
--- a/arch/arm/boot/dts/imx6qdl-sabresd-btwifi.dtsi
+++ b/arch/arm/boot/dts/imx6qdl-sabresd-btwifi.dtsi
@@ -14,6 +14,8 @@
  * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected via ribbon cable (J13 connector
  * on board).
  * ==> Hardware modification is required. Refer to schematic.
+ *
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration. 
  */
 
 / {
@@ -22,23 +24,17 @@
 		status = "disabled";
 	};
 
-	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio4 7 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio1 2 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
 	};
 
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		gpios = <&gpio4 6 0>; /* WL_HOST_WAKE */
-		wlreg_on-supply = <&wlreg_on>;
-	};
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio4 7 GPIO_ACTIVE_LOW>;
+ 	};
 };
 
 &ecspi1 {
@@ -47,6 +43,12 @@
 
 &iomuxc {
 	imx6qdl-sabresd-murata-v2 {
+		pinctrl_btreg: btreggrp {
+			fsl,pins = <
+				MX6QDL_PAD_GPIO_2__GPIO1_IO02 0x1b0b0  /* BT_REG_ON */
+			>;
+		};
+
 		/* add MUXing entry for SD2 4-bit interface and configure control pins */
 		pinctrl_wifi: wifigrp {
 			fsl,pins = <
@@ -63,10 +65,18 @@
 	};
 };
 
+/* re-defining gpio_leds removes conflict with BT_REG_ON */
+&pinctrl_gpio_leds {
+	fsl,pins = <
+	>;
+};
+
 &uart5 {
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_uart5_1>;
+	pinctrl-0 = <&pinctrl_uart5_1
+		     &pinctrl_btreg>;
 	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
 	status = "okay";
 	/* for DTE mode, add below change */
 	/* fsl,dte-mode; */
@@ -74,12 +84,23 @@
 };
 
 &usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_wifi>;
 	bus-width = <4>;
 	no-1-8-v;
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio4>;
+		interrupts = <6 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO4_IO6 active high. Change to IRQ_TYPE_EDGE_RISING for 1BW and SN8000. */
+		interrupt-names = "host-wake";
+	};
 };
diff --git a/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts
new file mode 100644
index 00000000..b08baf8
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6qp-sabresd-btwifi-m2.dts"
+#include "imx6q-sabresd-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts
new file mode 100644
index 00000000..1c4a097
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qp-sabresd-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6qp-sabresd.dts"
+#include "imx6qdl-sabresd-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts
new file mode 100644
index 00000000..95c01df
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sl-evk-btwifi-m2.dts
@@ -0,0 +1,103 @@
+//#include "imx6sl-evk-btwifi.dts"
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/* NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into SD1
+ * slot using Murata i.MX InterConnect Ver 1.0 Adapter AND wiring in control
+ * signals with SD Card Extender on SD3 slot.
+ * Bluetooth UART connect via SD1 EMMC/MMC Plus pinout.
+ * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected from SD Card Extender on SD3
+ * slot.
+ *
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration.
+ */
+#include "imx6sl-evk.dts"
+
+/ {
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 17 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 16 GPIO_ACTIVE_LOW>;
+ 	};
+};
+
+&iomuxc {
+	imx6sl-evk-murata-v1_sdext {
+		/* Only MUX SD1_DAT0..3 lines so UART4 can be MUXed on higher data lines. */
+		pinctrl_btreg: btreggrp {
+			fsl,pins = <
+				MX6SL_PAD_SD3_DAT3__GPIO5_IO17          0x13069 /* BT_REG_ON */
+			>;
+		};
+		pinctrl_wifi: wifigrp {
+			fsl,pins = <
+				MX6SL_PAD_SD1_CMD__SD1_CMD		0x17059
+				MX6SL_PAD_SD1_CLK__SD1_CLK		0x10059
+				MX6SL_PAD_SD1_DAT0__SD1_DATA0		0x17059
+				MX6SL_PAD_SD1_DAT1__SD1_DATA1		0x17059
+				MX6SL_PAD_SD1_DAT2__SD1_DATA2		0x17059
+				MX6SL_PAD_SD1_DAT3__SD1_DATA3		0x17059
+				MX6SL_PAD_SD3_DAT1__GPIO5_IO20		0x13069 /* WL_HOST_WAKE */
+				MX6SL_PAD_SD3_DAT2__GPIO5_IO16		0x13069 /* WL_REG_ON */
+			>;
+		};
+
+		pinctrl_usdhc3_1: usdhc3grp-1 {
+			fsl,pins = <
+				MX6SL_PAD_SD3_CMD__SD3_CMD		0x17059
+				MX6SL_PAD_SD3_CLK__SD3_CLK		0x10059
+				MX6SL_PAD_SD3_DAT0__SD3_DATA0		0x17059
+			>;
+		};
+	};
+};
+
+/* Murata: declare UART4 interface for Bluetooth. */
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart4_1
+		     &pinctrl_btreg>;
+	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode; */
+	/* pinctrl-0 = <&pinctrl_uart4dte_1>; */
+};
+
+&usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_wifi>;
+	bus-width = <4>;
+	no-1-8-v;
+	non-removable;
+	cap-power-off-card;
+	pm-ignore-notify;
+	wifi-host;
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
+
+&usdhc3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usdhc3_1>;
+	bus-width = <1>;
+	no-1-8-v;
+};
diff --git a/arch/arm/boot/dts/imx6sl-evk-btwifi.dts b/arch/arm/boot/dts/imx6sl-evk-btwifi.dts
index b740c81..bb64aab2 100644
--- a/arch/arm/boot/dts/imx6sl-evk-btwifi.dts
+++ b/arch/arm/boot/dts/imx6sl-evk-btwifi.dts
@@ -12,32 +12,33 @@
  * Bluetooth UART connect via SD1 EMMC/MMC Plus pinout.
  * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE are connected from SD Card Extender on SD3
  * slot.
+ *
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration.
  */
 #include "imx6sl-evk.dts"
 
 / {
-	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio5 16 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio5 17 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
 	};
 
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		gpios = <&gpio5 20 0>; /* WL_HOST_WAKE */
-		wlreg_on-supply = <&wlreg_on>;
-	};
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 16 GPIO_ACTIVE_LOW>;
+ 	};
 };
 
 &iomuxc {
 	imx6sl-evk-murata-v1_sdext {
 		/* Only MUX SD1_DAT0..3 lines so UART4 can be MUXed on higher data lines. */
+		pinctrl_btreg: btreggrp {
+			fsl,pins = <
+				MX6SL_PAD_SD3_DAT3__GPIO5_IO17          0x13069 /* BT_REG_ON */
+			>;
+		};
 		pinctrl_wifi: wifigrp {
 			fsl,pins = <
 				MX6SL_PAD_SD1_CMD__SD1_CMD		0x17059
@@ -61,15 +62,39 @@
 	};
 };
 
+/* Murata: declare UART4 interface for Bluetooth. */
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart4_1
+		     &pinctrl_btreg>;
+	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode; */
+	/* pinctrl-0 = <&pinctrl_uart4dte_1>; */
+};
+
 &usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_wifi>;
 	bus-width = <4>;
 	no-1-8-v;
 	non-removable;
-	cd-post;
+	cap-power-off-card;
 	pm-ignore-notify;
 	wifi-host;
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio5>;
+		interrupts = <20 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO5_IO20 active high. Change to IRQ_TYPE_EDGE_RISING for 1BW and SN8000. */
+		interrupt-names = "host-wake";
+	};
 };
 
 &usdhc3 {
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts
new file mode 100644
index 00000000..0a19a8d
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6sx-sdb-btwifi-m2.dts"
+#include "imx6sx-sdb-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..452bbef
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2-oob.dtsi
@@ -0,0 +1,18 @@
+&iomuxc {
+	imx6sx-sdb-murata-v1_sdext {
+		pinctrl_wifi_oob: pinctrl_wifi {
+			fsl,pins = <
+				// Murata Module control signals -- not connected (R327 and need to disable ENET1 CLK)
+				MX6SX_PAD_ENET1_RX_CLK__GPIO2_IO_4	0x0b000 /* WL_HOST_WAKE */
+			>;
+		};
+	};
+};
+&fec1 {
+	status = "disabled";
+};
+&brcmf{
+	interrupt-parent = <&gpio2>;
+	interrupts = <4 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts
new file mode 100644
index 00000000..3038da0
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dts
@@ -0,0 +1,2 @@
+#include "imx6sx-sdb.dts"
+#include "imx6sx-sdb-btwifi-m2.dtsi"
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi
new file mode 100644
index 00000000..4c7d8d7
--- /dev/null
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi-m2.dtsi
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+/*
+ * NOTE: This DTS file is written for plugging in Murata Wi-Fi/BT EVK into SD3
+ * slot using Murata i.MX InterConnect Ver 1.0 Adapter AND SD Card Extender on
+ * SD2 slot. Bluetooth UART connects via SD3 EMMC/MMC Plus pinout.
+ * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE connect via SD Card Extender.
+ */
+
+/ {
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio4 10 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+//		reset-gpios = <&gpio2 4 GPIO_ACTIVE_LOW>;  /* currently not implemented */
+	};
+};
+
+&iomuxc {
+	imx6sx-sdb-murata-wifibt {
+		pinctrl_bt: btgrp {
+			fsl,pins = <
+				MX6SX_PAD_NAND_DATA06__GPIO4_IO_10	0x13069 /* BT_REG_ON */
+			>;
+		};
+
+		pinctrl_wifi: wifigrp {
+			fsl,pins = <
+				/* Murata Module control signals -- not connected (R327 and need to disable ENET1 CLK) */
+//				MX6SX_PAD_ENET1_RX_CLK__GPIO2_IO_4	0x13069 /* WL_REG_ON */
+			>;
+		};
+	};
+};
+
+&uart5 { /* for bluetooth */
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart5 &pinctrl_bt>;
+	uart-has-rtscts;
+	resets = <&modem_reset>;
+	status = "okay";
+	/* for DTE mode, add below change */
+	/* fsl,dte-mode;*/
+	/* pinctrl-0 = <&pinctrl_uart5dte_1>; */
+};
+
+
+&usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usdhc2 &pinctrl_wifi>;
+	bus-width = <4>;
+	no-1-8-v;	/* force 3.3V VIO */
+	non-removable;
+	mmc-pwrseq = <&usdhc2_pwrseq>;
+	pm-ignore-notify;
+	cap-power-off-card;
+	status = "okay";
+
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
diff --git a/arch/arm/boot/dts/imx6sx-sdb-btwifi.dts b/arch/arm/boot/dts/imx6sx-sdb-btwifi.dts
index ba5512e..8dc07bf 100644
--- a/arch/arm/boot/dts/imx6sx-sdb-btwifi.dts
+++ b/arch/arm/boot/dts/imx6sx-sdb-btwifi.dts
@@ -11,28 +11,23 @@
  * slot using Murata i.MX InterConnect Ver 1.0 Adapter AND SD Card Extender on
  * SD2 slot. Bluetooth UART connects via SD3 EMMC/MMC Plus pinout.
  * WL_REG_ON/BT_REG_ON/WL_HOST_WAKE connect via SD Card Extender.
+ * 
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ configuration.
  */
 
 #include "imx6sx-sdb.dts"
 
 / {
-	regulators {
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio6 10 0>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio6 11 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
 	};
 
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		/* WL_HOST_WAKE: SD2_DAT1 (gpio6 9) */
-		gpios = <&gpio6 9 0>;
-		wlreg_on-supply = <&wlreg_on>;
+	usdhc3_pwrseq: usdhc3_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio6 10 GPIO_ACTIVE_LOW>;
 	};
 };
 
@@ -86,6 +81,7 @@
 	pinctrl-0 = <&pinctrl_uart3
 		     &pinctrl_bt>;
 	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
 	status = "okay";
 };
 
@@ -100,13 +96,23 @@
 };
 
 &usdhc3 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_wifi>;
 	bus-width = <4>;
 	no-1-8-v;	/* force 3.3V VIO */
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host;	/* pull in card detect mechanism for BCMDHD driver */
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc3_pwrseq>;
 	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio6>;
+		interrupts = <9 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO6_IO9 active high. Change to IRQ_TYPE_EDGE_RISING for 1BW and SN8000. */
+		interrupt-names = "host-wake";
+	};
 };
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts
new file mode 100644
index 00000000..4868d39
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob-uhs.dts
@@ -0,0 +1,19 @@
+#include "imx6ul-14x14-evk-btwifi-m2-uhs.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
+
+&fec2 {
+	status="disabled";
+};
+
+&fec1 {
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+		};
+	};
+};
+
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts
new file mode 100644
index 00000000..5826e96
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-oob.dts
@@ -0,0 +1,2 @@
+#include "imx6ul-14x14-evk-btwifi-m2.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts
new file mode 100644
index 00000000..fd55b16
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2-uhs.dts
@@ -0,0 +1,11 @@
+#include "imx6ul-14x14-evk-btwifi-m2.dts"
+
+&usdhc1 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
+	/delete-property/ no-1-8-v;
+};
+
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts
new file mode 100644
index 00000000..f8646c2
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk-btwifi-m2.dts
@@ -0,0 +1,64 @@
+#include "imx6ul-14x14-evk-btwifi.dts"
+
+/ {
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio_spi 4 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 1 GPIO_ACTIVE_LOW>; /* WL REG ON */
+		post-power-on-delay-ms = <120>;  /* Murata - add delay before accessing SDIO bus */
+	};
+};
+
+&wlreg_on {
+	status = "disabled";
+};
+
+&bcmdhd_wlan_0 {
+	status = "disabled";
+};
+
+&pinctrl_hog_1 {
+        fsl,pins = <
+            	MX6UL_PAD_GPIO1_IO05__GPIO1_IO05	0x17059 /* SD1 VSELECT */
+        >;
+};
+
+&uart2 {
+	pinctrl-0 = <&pinctrl_uart2>;
+	resets = <&modem_reset>;
+};
+
+&usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	fsl,tuning-start-tap = <20>;
+	fsl,tuning-step = <2>;
+	no-1-8-v;
+	non-removable;
+	/delete-property/ cd-post;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	/delete-property/ wifi-host;
+
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
+
+&gpio_spi {
+	/* Murata: modify default setting so that BT_nPWD/BT_REG_ON
+	 * is low (0V) during kernel boot.
+	 */
+	registers-default = /bits/ 8 <0x47>;	/* BT_REG_ON */
+};
+
diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk.dts b/arch/arm/boot/dts/imx6ul-14x14-evk.dts
index d643665..3937aa6 100644
--- a/arch/arm/boot/dts/imx6ul-14x14-evk.dts
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk.dts
@@ -168,12 +168,24 @@
 	};
 };
 
+/* Murata: need to redefine FEC1 and FEC2 constructs to keep MDIO declarations independent. 
+ * Needed by OOB IRQ functionality - see imx6ul-evk-btwifi-oob.dtsi 
+ */
 &fec1 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_enet1>;
 	phy-mode = "rmii";
 	phy-handle = <&ethphy0>;
 	status = "okay";
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+		};
+	};
 };
 
 &fec2 {
@@ -187,11 +199,6 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		ethphy0: ethernet-phy@2 {
-			compatible = "ethernet-phy-ieee802.3-c22";
-			reg = <2>;
-		};
-
 		ethphy1: ethernet-phy@1 {
 			compatible = "ethernet-phy-ieee802.3-c22";
 			reg = <1>;
diff --git a/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi b/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi
new file mode 100644
index 00000000..66e55dd
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ul-evk-btwifi-m2-oob.dtsi
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2017 NXP
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+&pinctrl_wifi {
+	fsl,pins = <
+		/* MUXing for WL_HOST_WAKE */
+		MX6UL_PAD_ENET2_RX_ER__GPIO2_IO15  0x0b000  /* input: 100K Pull-up */
+	>;
+};
+
+/*
+ * For WL_HOST_WAKE (OOB_IRQ) to function correctly, we must disable
+ * the secondary ethernet port (FEC2). Hardware re-work is to remove
+ * R1633 and populate R1704 with 0 Ohm resistor.
+ * Refer to Murata Hardware Reference Manual for more details.
+ */
+&fec2 {
+	status = "disabled";
+};
+
+&brcmf {
+	interrupt-parent = <&gpio2>;
+	interrupts = <15 IRQ_TYPE_LEVEL_LOW>;
+	interrupt-names = "host-wake";
+};
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts
new file mode 100644
index 00000000..70d4b06
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob-uhs.dts
@@ -0,0 +1,19 @@
+#include "imx6ull-14x14-evk-btwifi-m2-uhs.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
+
+&fec2 {
+	status="disabled";
+};
+
+&fec1 {
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+		};
+	};
+};
+
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts
new file mode 100644
index 00000000..6b77b80
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-oob.dts
@@ -0,0 +1,3 @@
+#include "imx6ull-14x14-evk-btwifi-m2.dts"
+#include "imx6ul-evk-btwifi-m2-oob.dtsi"
+
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts
new file mode 100644
index 00000000..e6372f6b
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2-uhs.dts
@@ -0,0 +1,11 @@
+#include "imx6ull-14x14-evk-btwifi-m2.dts"
+
+&usdhc1 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz"; /* Murata -- add 100 and 200 Mhz entries for UHS mode */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	pinctrl-1 = <&pinctrl_usdhc1_100mhz &pinctrl_wifi>;
+	pinctrl-2 = <&pinctrl_usdhc1_200mhz &pinctrl_wifi>;
+	max-frequency = <132000000>; /* Murata - 6ULL MAX with Adapter; 6UL = 100 MHz MAX */
+	/delete-property/ no-1-8-v;
+};
+
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts
new file mode 100644
index 00000000..2de4779
--- /dev/null
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk-btwifi-m2.dts
@@ -0,0 +1,61 @@
+#include "imx6ull-14x14-evk-btwifi.dts"
+
+/ {
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio_spi 4 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc1_pwrseq: usdhc1_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio5 1 GPIO_ACTIVE_LOW>; /* WL REG ON */
+		post-power-on-delay-ms = <120>;  /* Murata - add delay before accessing SDIO bus */
+	};
+};
+
+&wlreg_on {
+	status = "disabled";
+};
+
+&bcmdhd_wlan_0 {
+	status = "disabled";
+};
+
+&pinctrl_hog_1 {
+        fsl,pins = <
+            	MX6UL_PAD_GPIO1_IO05__GPIO1_IO05	0x17059 /* SD1 VSELECT */
+        >;
+};
+
+&uart2 {
+	pinctrl-0 = <&pinctrl_uart2>;
+	resets = <&modem_reset>;
+};
+
+&usdhc1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-assert-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>; /* Force 1.8V VIO */
+	pinctrl-0 = <&pinctrl_usdhc1 &pinctrl_wifi>;
+	non-removable;
+	/delete-property/ cd-post;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	/delete-property/ wifi-host;
+
+	mmc-pwrseq = <&usdhc1_pwrseq>;
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
+
+&gpio_spi {
+	/* Murata: modify default setting so that BT_nPWD/BT_REG_ON
+	 * is low (0V) during kernel boot.
+	 */
+	registers-default = /bits/ 8 <0x47>;	/* BT_REG_ON */
+};
+
diff --git a/arch/arm/boot/dts/imx6ull-14x14-evk.dts b/arch/arm/boot/dts/imx6ull-14x14-evk.dts
index b0fb913..48c815d 100644
--- a/arch/arm/boot/dts/imx6ull-14x14-evk.dts
+++ b/arch/arm/boot/dts/imx6ull-14x14-evk.dts
@@ -168,12 +168,24 @@
 	};
 };
 
+/* Murata: need to redefine FEC1 and FEC2 constructs to keep MDIO declarations independent. 
+ * Needed by OOB IRQ functionality - see imx6ul-evk-btwifi-oob.dtsi 
+ */
 &fec1 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_enet1>;
 	phy-mode = "rmii";
 	phy-handle = <&ethphy0>;
 	status = "okay";
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethphy0: ethernet-phy@2 {
+			compatible = "ethernet-phy-ieee802.3-c22";
+			reg = <2>;
+		};
+	};
 };
 
 &fec2 {
@@ -187,11 +199,6 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		ethphy0: ethernet-phy@2 {
-			compatible = "ethernet-phy-ieee802.3-c22";
-			reg = <2>;
-		};
-
 		ethphy1: ethernet-phy@1 {
 			compatible = "ethernet-phy-ieee802.3-c22";
 			reg = <1>;
diff --git a/arch/arm/boot/dts/imx7d-sdb.dts b/arch/arm/boot/dts/imx7d-sdb.dts
index 8f03224..2f0ecf36 100644
--- a/arch/arm/boot/dts/imx7d-sdb.dts
+++ b/arch/arm/boot/dts/imx7d-sdb.dts
@@ -4,6 +4,8 @@
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 as
  * published by the Free Software Foundation.
+ * 
+ * Murata: modified to support WLAN FMAC driver with OOB IRQ. 
  */
 
 /dts-v1/;
@@ -19,6 +21,18 @@
 		reg = <0x80000000 0x80000000>;
 	};
 
+	modem_reset: modem-reset {
+		compatible = "gpio-reset";
+		reset-gpios = <&gpio4 23 GPIO_ACTIVE_LOW>;
+		reset-delay-us = <1000>;
+		#reset-cells = <0>;
+	};
+
+	usdhc2_pwrseq: usdhc2_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&gpio4 21 GPIO_ACTIVE_LOW>;
+	};
+
 	backlight {
 		compatible = "pwm-backlight";
 		pwms = <&pwm1 0 5000000>;
@@ -27,12 +41,6 @@
 		status = "okay";
 	};
 
-	bcmdhd_wlan_0: bcmdhd_wlan@0 {
-		compatible = "android,bcmdhd_wlan";
-		gpios = <&gpio4 20 0>;		/* WL_HOST_WAKE */
-		wlreg_on-supply = <&wlreg_on>;
-	};
-
 	pxp_v4l2_out {
 		compatible = "fsl,imx7d-pxp-v4l2", "fsl,imx6sx-pxp-v4l2", "fsl,imx6sl-pxp-v4l2";
 		status = "okay";
@@ -145,16 +153,6 @@
 			regulator-min-microvolt = <1800000>;
 			regulator-max-microvolt = <1800000>;
 		};
-
-		wlreg_on: fixedregulator@100 {
-			compatible = "regulator-fixed";
-			regulator-min-microvolt = <5000000>;
-			regulator-max-microvolt = <5000000>;
-			regulator-name = "wlreg_on";
-			gpio = <&gpio4 21 GPIO_ACTIVE_HIGH>;
-			startup-delay-us = <100>;
-			enable-active-high;
-		};
 	};
 };
 
@@ -551,7 +549,6 @@
 	imx7d-sdb {
 		pinctrl_hog_1: hoggrp-1 {
 			fsl,pins = <
-				MX7D_PAD_ECSPI2_SS0__GPIO4_IO23	0x19 /* BT_REG_ON */
 				MX7D_PAD_EPDC_BDR0__GPIO2_IO28	0x59 /* headphone detect */
 			>;
 		};
@@ -841,6 +838,7 @@
 				MX7D_PAD_ECSPI1_SCLK__UART6_DCE_RX 0x79
 				MX7D_PAD_ECSPI1_SS0__UART6_DCE_CTS 0x79
 				MX7D_PAD_ECSPI1_MISO__UART6_DCE_RTS 0x79
+				MX7D_PAD_ECSPI2_SS0__GPIO4_IO23	0x19  /* BT_REG_ON */
 			>;
 		};
 
@@ -913,6 +911,8 @@
 				MX7D_PAD_SD2_DATA1__SD2_DATA1   0x5a
 				MX7D_PAD_SD2_DATA2__SD2_DATA2   0x5a
 				MX7D_PAD_SD2_DATA3__SD2_DATA3   0x5a
+				MX7D_PAD_ECSPI2_MOSI__GPIO4_IO21	0x19 /* WL_REG_ON */
+				MX7D_PAD_ECSPI2_SCLK__GPIO4_IO20	0x19 /* WL_HOST_WAKE */
 			>;
 		};
 
@@ -924,6 +924,8 @@
 				MX7D_PAD_SD2_DATA1__SD2_DATA1   0x5b
 				MX7D_PAD_SD2_DATA2__SD2_DATA2   0x5b
 				MX7D_PAD_SD2_DATA3__SD2_DATA3   0x5b
+				MX7D_PAD_ECSPI2_MOSI__GPIO4_IO21	0x19 /* WL_REG_ON */
+				MX7D_PAD_ECSPI2_SCLK__GPIO4_IO20	0x19 /* WL_HOST_WAKE */
 			>;
 		};
 
@@ -1125,6 +1127,7 @@
 	assigned-clocks = <&clks IMX7D_UART6_ROOT_SRC>;
 	assigned-clock-parents = <&clks IMX7D_PLL_SYS_MAIN_240M_CLK>;
 	fsl,uart-has-rtscts;
+	resets = <&modem_reset>;
 	status = "okay";
 };
 
@@ -1156,17 +1159,25 @@
 };
 
 &usdhc2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	pinctrl-names = "default", "state_100mhz", "state_200mhz";
 	pinctrl-0 = <&pinctrl_usdhc2>;
 	pinctrl-1 = <&pinctrl_usdhc2_100mhz>;
 	pinctrl-2 = <&pinctrl_usdhc2_200mhz>;
-	enable-sdio-wakeup;
-	keep-power-in-suspend;
 	non-removable;
-	cd-post;
-	pm-ignore-notify;
-	wifi-host;
+	cap-power-off-card;
+	pm-ignore-notify; 
+	wifi-host; 
+	mmc-pwrseq = <&usdhc2_pwrseq>;
 	status = "okay";
+	brcmf: bcrmf@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+		interrupt-parent = <&gpio4>;
+		interrupts = <20 IRQ_TYPE_LEVEL_HIGH>;  /* WL_HOST_WAKE = GPIO4_IO20 active high */
+		interrupt-names = "host-wake";
+	};
 };
 
 &usdhc3 {
